# Methods {#methods}

## Plant Material and Traits Collected

From 2011 to 2014, cranberry uprights bearing fruit were delivered from samples taken at the New Jersey Agricultural
Experiment Station (NJAES).  These samples were derived from two full-sib interrelated breeding populations, CNJ02
($n=169$) and CNJ04 ($n=72$).  The CNJ02 population included progeny from a cross between maternal parent, CNJ97-105
(_Mullica Queen_), and paternal parent, NJS98-23 (_Crimson Queen_); the smaller CNJ04 population included progeny from a
reciprocal cross between CNJ97-105 (_Mullica Queen_) and _Stevens_.

Cranberry traits have customarily been measured on a per-upright basis and the traits collected on CNJ02 and CNJ04
represent a set of traits traditionally believed to be important to cranberries (Vorsa and Zalapa personal
communication).  In this instance, the following traits were manually gathered per upright, 10 uprights per genotype:
number of pedicels (NP), number of berries (NB), number of aborted flowers (NFA), total berry weight (TBM), upright
length (UL), and evidence of rebud (RB).  Moreover, the following traits were collected on the largest fruit per
upright: length (BL), width (BW), weight (BM), calyx diameter (CD), calyx lobe characteristics (CLC), shape
characteristics (S), amount of fruit bloom or bloom rating (BR), and number of developed seeds (NS).  A summary of
traits collected is displayed in Table \@ref(tab:trait-tab-kable).

```{r trait-tab-setup, echo=FALSE, message=FALSE, include=TRUE, results='asis'}
trait_tab <- readr::read_delim("../Data/publication/tables/trait_description_table.csv", delim=",", escape_backslash=TRUE, na=c("NA")) 
trait_tab_cat <- trait_tab %>% 
    dplyr::select(Trait,Abbreviation,Category,Units,Description) %>%
    group_by(Category) %>%
    mutate(category_repeat=TRUE) %>%
    do(tibble::add_row(.,
                       Trait=.$Category[1], #Replace trait with category for making a more compact table
                       Abbreviation="",
                       Category=.$Category[1],
                       Units="",
                       Description="",
                       category_repeat=FALSE,
                       .before=1 #Inject at top of tibble
                       )) %>%
    ungroup()
```

```{r trait-tab-kable, echo=FALSE, message=FALSE, include=TRUE, results='asis', tab.cap="Summary of traits collected from cranberry (Vaccinium macrocarpon Ait.) populations CNJ02 and CNJ04 in years 2011-2014."}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')
trait_foot <- "Upright Traits were measured on 10 fruiting-uprights. Largest berry traits represent phenotypic values of the largest berry on a given fruiting upright. Plot traits were collected by harvesting berries from a 0.09 m2 plot sample and running appropriate assays onsite at Marucci Center. Upright Traits and Largest Berry Traits were collected at Rutgers and assayed at UW-Madison. Trait abbreviations are used to label subsequent figures.  Berry Shape Chimera Parameters were derived from canonical representations of berry shapes.  Expressions between parentheses indicate trait units or categorical values."
abbreviation_foot <- "Trait abbreviations are used to label subsequent figures."

if ( is_word_output() ) {
ftrait_tab_cat <-   
    dplyr::select(trait_tab_cat,c(Trait,Abbreviation,Units,Description)) %>%
                    mutate(Units=gsub("^\\$","",Units)) %>%
                    mutate(Units=gsub("\\$$","",Units)) %>%
                    flextable(col_keys=c("Trait","Abbreviation","Description")) %>%
                    set_header_labels(Abbreviation="Abbv") %>%
                    fontsize(part="body",
                            size=11) %>%
                    hline( i = which(trait_tab_cat$category_repeat == TRUE), part="body", border=fp_border_default(color="#DDDDDD", width=0.25)) %>%
                    hline( i = (which(trait_tab_cat$category_repeat == FALSE)-1)[-1], part="body") %>%
                    padding(j = "Trait",
                            padding.left = 12) %>%
                    bold(i = ~ trait_tab_cat$category_repeat == FALSE, 
                             j = "Trait", 
                             part = "body") %>% 
                    fontsize(i = ~ trait_tab_cat$category_repeat == FALSE, 
                             j = "Trait", 
                             part = "body",
                             size = 12) %>%
                    padding(i = ~ trait_tab_cat$category_repeat == FALSE, 
                            j = "Trait", 
                            part="body",
                            padding.left=0) %>%
                    mk_par(i = (which(!is.empty(trait_tab_cat$Units))), j = "Trait", part="body", value=as_paragraph(Trait,"\n(",as_equation(Units),")")) %>%
                    fontsize(part="header",
                             size = 12) %>%
                    bold(i = 1,
                         part="header") %>%
                    flextable::footnote(i = 1, 
                            j = c("Trait", "Abbreviation"),
                            value = as_paragraph(c(trait_foot, abbreviation_foot)),
                            ref_symbols = c("a","b"),
                            part = "header") %>%
                    fontsize(part="footer",
                             size=10)
    flextable::save_as_docx(ftrait_tab_cat, 
                            path="../Data/publication/tables/categorical_traits.docx", 
                            pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                            align="center")
} else {
    ftrait_tab_cat <- rename(trait_tab_cat,'Category[note]'=Category,'Abbreviation[note]'=Abbreviation) %>%
        mutate(Units=ifelse(Units != "",paste0("$",Units,"$"),Units)) %>%
        kable(  booktabs=TRUE, 
                escape = FALSE,
                caption=caption) %>%
        column_spec(1, bold=TRUE) %>%
        column_spec(3, width="2cm") %>%
        column_spec(5, width="5cm") %>%
        row_spec(row=(which(trait_tab_cat["Category"] != "")-1)[-1], hline_after = TRUE) %>%
        kable_paper("striped", font_size=table.font_size, full_width=FALSE) %>%
        kable_styling(latex_options = "scale_down") %>%
        add_footnote(c(caption_ext,
                "Trait abbreviations are used to label subsequent figures."))
}
print_table(ftrait_tab_cat, clabel, caption) #Render
```

```{r cross-study-trait-tab-setup, echo=FALSE, message=FALSE, include=TRUE, results='asis'}
cross_study_trait_tab <- readr::read_delim("../Data/publication/tables/crossstudy_trait_description_table.csv", delim=",", escape_backslash=TRUE, na=c("NA")) 
cross_study_trait_tab_study <- cross_study_trait_tab %>% 
    group_by(Study) %>%
    mutate(study_repeat=TRUE) %>%
    do(tibble::add_row(.,
                       Study=.$Study[1],
                       Trait=.$Study[1], #Replace trait with category for making a more compact table
                       Abbreviation="",
                       Units="",
                       Description="",
                       study_repeat=FALSE,
                       .before=1 #Inject at top of tibble
                       )) %>%
    ungroup()
```

```{r cross-study-trait-tab, echo=FALSE, message=FALSE, include=TRUE, results='asis', tab.cap="Abbreviations and descriptions for relevant traits referenced from Diaz-Garcia et al. (2018a), Diaz-Garcia et al. (2018b), and Schlautman et al. (2015)."}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if ( is_word_output() ) {
ftrait_cross_study_tab_study <- cross_study_trait_tab_study %>%
                    mutate(Units=gsub("^\\$","",Units)) %>%
                    mutate(Units=gsub("\\$$","",Units)) %>%
                    flextable(col_keys=c("Trait","Abbreviation","Description")) %>%
                    set_header_labels(Abbreviation="Abbv") %>%
                    fontsize(part="body",
                            size=11) %>%
                    hline( i = which(cross_study_trait_tab_study$study_repeat == TRUE), part="body", border=fp_border_default(color="#DDDDDD", width=0.25)) %>%
                    hline( i = (which(cross_study_trait_tab_study$study_repeat == FALSE)-1)[-1], part="body") %>%
                    bold( i = which(cross_study_trait_tab_study$study_repeat == FALSE), part="body") %>%
                    italic( i = which(cross_study_trait_tab_study$study_repeat == FALSE), part="body") %>%
                    flextable::footnote(i = 1, 
                            j = c("Trait"),
                            value = as_paragraph("Bold italic text indicates the study of origin.  Entries between parentheses indicate the units of measurement."),
                            ref_symbols = c("a"),
                            part = "header") %>%
                    padding(j = "Trait",
                            padding.left = 12) %>%
                    fontsize(i = ~ study_repeat == FALSE, 
                             j = ~ Trait, 
                             part = "body",
                             size = 12) %>%
                    padding(i = ~ study_repeat == FALSE, 
                            j = ~ Trait, 
                            part="body" ,
                            padding.left=0) %>%
                    fontsize(part="header",
                             size = 12) %>%
                    bold(part="header") %>%
                    mk_par(i = (which(!is.empty(cross_study_trait_tab_study$Units))), 
                           j = ~ Trait, 
                           part="body", 
                           value=as_paragraph(Trait,"\n(",as_equation(Units),")")) %>%
                    fontsize(part="footer",
                             size=10)

    flextable::save_as_docx(ftrait_cross_study_tab_study,
                            path="../Data/publication/tables/cross_study_traits.docx", 
                            pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                            align="center")
} else {
}
```

```{r cross-study-trait-groups-setup, echo=FALSE, message=FALSE, include=TRUE, results='asis'}
cross_study_trait_groups_tab_orig <- readr::read_delim("../Data/publication/tables/crossstudy_trait_groups_table.csv", delim=",", escape_backslash=TRUE, na=c("NA")) 
cross_study_trait_groups_tab <- cross_study_trait_groups_tab_orig %>% 
    group_by(Set) %>%
    mutate(set_repeat=TRUE) %>%
    do(tibble::add_row(.,
                       Set=.$Set[1],
                       Trait_Groups=.$Set[1], #Replace trait groups with set
                       Studies=.$Studies[1],
                       Populations=.$Populations[1],
                       Filter="",
                       set_repeat=FALSE,
                       .before=1 #Inject at top of tibble
                       )) %>%
    ungroup() %>%
    mutate(Set=ifelse(set_repeat,"",Set),
           Studies=ifelse(set_repeat,"",Studies),
           Populations=ifelse(set_repeat,"",Populations))
```

```{r cross-study-trait-groups-tab, echo=FALSE, message=FALSE, include=TRUE, results='asis', tab.cap="Trait groupings for each set of collocated QTL. Trait groups were selected based on correlated clusters or similar categories of traits. The Filter column details the constraints applied to the synthetic metaQTL that are displayed in the linkage maps. ▁(R^(2 ) ) is the mean percent marker variance explained across all composite QTL."}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if ( is_word_output() ) {
wf <- which(cross_study_trait_groups_tab$set_repeat == FALSE)
wt <- which(cross_study_trait_groups_tab$set_repeat == TRUE)
fcross_study_trait_groups_tab <- cross_study_trait_groups_tab %>%
                    mutate(Filter=gsub("^\\$","",Filter)) %>%
                    mutate(Filter=gsub("\\$$","",Filter)) %>%
                    flextable(col_keys=c("Trait_Groups","Studies","Populations","Filter")) %>%
                    set_header_labels(Trait_Groups="Trait Groups") %>%
                    fontsize(part="body",
                             size=11) %>%
                    hline(i = wt[1:length(wt)-1], 
                          j = ~Trait_Groups,
                          part="body", 
                          border=fp_border_default(color="#DDDDDD", width=0.25)) %>%
                    hline(i = (wf-1)[-1], 
                          part="body") %>%
                    bold(i = wf, 
                         j = ~Trait_Groups,
                         part="body") %>%
                    italic(i = wf, 
                           j = ~Trait_Groups,
                           part="body") %>%
                    flextable::footnote(i = 1, 
                            j = c("Trait_Groups"),
                            value = as_paragraph("Bold entries designate the separate sets used in colocation QTL analysis."),
                            ref_symbols = c("a"),
                            part = "header") %>%
                    padding(j = ~ Trait_Groups,
                            padding.left = 12) %>%
                    fontsize(i = wf, 
                             j = ~ Trait_Groups, 
                             part = "body",
                             size = 12) %>%
                    padding(i = wf, 
                            j = ~ Trait_Groups, 
                            part="body" ,
                            padding.left=0) %>%
                    fontsize(part="header",
                             size = 12) %>%
                    bold(part="header") %>%
                    mk_par(j = ~ Filter, 
                           part="body", 
                           value=as_paragraph(as_equation(Filter))) %>%
                    fontsize(part="footer",
                             size=10)
                    
    flextable::save_as_docx(fcross_study_trait_groups_tab,
                            path="../Data/publication/tables/cross_study_trait_groups.docx", 
                            pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                            align="center")
} else {
}
```


## Trait Evaluation and Transformation

Except for largest berry shape, categorical traits were transformed to numeric, discrete values.  Higher values
represented more favorable characteristics.    The largest berry categorical to digital map is shown in
\@ref(tab:cat-trait-tab).

To quantify largest berry shape, scored categorical shapes were first mapped to digital image shapes by starting with the template images provided
in Figure \@ref(fig:berry-shape-template).  These images were first converted to binary images, and then normalized by
centering all binarized representations and scaling so that all shape templates have the same area.  Thereafter, a
chimeric berry representation was derived for each genotype-year by combining and thresholding the differential distance transformed
representations of each corresponding upright's normalized berry shape.  A distance transform representation of a binary
image is the euclidean distance of any foreground pixel to its nearest background pixel.  The differential distance
transformiation is the difference between the distance transformation of the normalized binary berry template image and
the same image inverted.  Positive values indicate foreground pixels, with higher values indicating the centroid of
binary images.  By combining the differential distance transforms and thresholding on positive values, a chimeric binary berry
image representation per genotype-year can be generated.  Shape descriptors can be subsequently calculated on this
representative berry, allowing for quantitative analysis of these traits for mapping.  This process is demonstrated in
Figure \@ref(fig:berry-chimera-table).

Digital image shape descriptors derived from the representative berry chimera are listed in Table \@ref(tab:trait-tab-kable).
Traits *UKUX* and *UKUY* are $log_{10}$
derivations of unsigned manhattan chain codes [@zalik2016].  In brief, a chain
code is a numerical representation describing the contour path of an object.  Unsigned manhattan chain codes are one of
many chain codes used to desribe contour shapes, and were chosen here for their highly compressible representation.
Other chimera shape descriptors outlined in Table \@ref(tab:trait-tab-kable) include *UKEC*, or chimera shape eccentricity,
*UKLvW*, or chimera length versus width, *UKTO*, chimera tortuosity, and *UKSO*, chimera solidity.  Eccentricity is a
mathematical descriptor for the curvature of an ellipse, with zero indicating a perfect circle and anything less than
one, but higher, indicating a higher length versus width.  Tortuosity describes the *'waviness'* of an object's contour,
and in this case, a slope chain code (SCC) method was applied to generate this tortuosity value [@Bribiesca2013].  Solidity
describes the shape density relative to it’s convex hull.  Lower solidity values indicate more *'waviness'* in the berry
contour, while higher values have smoother contour curvature.

All traits were curated by removing entries marked as rotten, and outliers were detected and trimmed by assessing residuals of each trait regressed by population, genotype, and year.  Additional observations were culled if 
they exceeded three standard deviation units from the mean under a Gaussian standard distribution.   Subsequent analysis and trait mapping was applied to all traits based the mean trait values across the ten sampled uprights, for each genotype-year.
From these upright means, Pearson correlation coefficients were calculated for all traits, averaged across months and all sampled years.  Correlation heatmap plots were generated using the 'corrplot' package, with statistical *p-values* calculated using the 
*cor_pmat()* function in the 'ggcorrplot' package [@corrplot2021;@ggcorrplot2019].  Additionally, correlation plots
sorted using the 'TSP' algorithm with the 'seriation' software package to highlight highly correlated (positive and negative) blocks of traits [@Hahsler2008; @Hahsler2021].


## Estimating Breeding Values and Heritability

Equation \@ref(eq:mm-year) shows the mixed model used in estimating BLUPs within years. The equation variables are defined as follows: $y$ = phenotype value, $μ$ = intercept (global mean of trait), 
$Z_g$ = genotype random-effect incidence matrix, $g$ = genotypic effects (BLUPs), $Z_r$ = row random effect incidence matrix, $r$ = row effect, $Z_c$ = column random effect incidence matrix, $c$ = column effect, 
$Z_s$ = 2D-spline random effect incidence matrix, $s$ = spline effect, $ε$ = residuals, $G$ = genotype variance-covariance matrix [@DelosCampos2015], $A$ = additive genomic relationship matrix [@Endelman2011], 
$σ_a^2$ = additive genomic variance, $I$ = identity matrix, $σ_r^2$ = row variance, $σ_c^2$ = column variance, $σ_s^2$ = 2D-spline variance, $σ_{ε}^2$ = residual error variance.

\begin{equation} 
    y = \mu + Z_{g}g + Z_{r}r + Z_{c}c + Z_{s}s + ε
    (\#eq:mm-year)
\end{equation} 

where $g \sim \mathcal{N}(0,G)$, $G = Aσ_a^2$, $r \sim \mathcal(0, Iσ_r^2)$, $c \sim \mathcal{N}(0,Iσ_c^2)$, $s \sim \mathcal{N}(0,Iσ_s^2)$, and $ε \sim \mathcal{N}(0,Iσ_{ε}^2)$.

Equation \@ref(eq:mm-year-ge) displays the across year mixed model used in estimating BLUPs.  Year is modeled as a fixed effect.   All symbols are the same as in Equation \@ref(eq:mm-year), 
but with the additional term $X_e$ = fixed-effect year incidence matrix, $e$ = year effect, $Z_{ge}$ = genotype-by-year random-effect incidence matrix, $ge$ = genotype-by-year effect, and $σ_{ge}^2$ = genotype-by-year variance.

\begin{equation} 
    y = \mu + X_{e}e + Z_{g}g + Z_{ge}ge + Z_{r}r + Z_{c}c + Z_{s}s + ε
    (\#eq:mm-year-ge)
\end{equation} 

where $y \in \{2011,2012,2013\}$, and $g_e \sim \mathcal{N}(0,σ_{ge}^2)$._


The `sommer` R package provides a powerful and flexible software library for specifying multiple variance-covariance structures of random-effects when using likelihood-based methods to fit mixed-models [@Covarrubias-Pazaran2016a].  These mixed-models
were originally proposed by Henderson, allowing the development of a statistical methodology that leverages animal-kinship relationships to better predict breeding values [@Henderson1975].  These methods have subsequently been adopted 
by plant breeders, having shown success in increasing the effectiveness of plant breeding.  With the expansion of genomic data acquisition and the development of dense marker maps, the use of these genetic maps can be used to model additive, 
dominance, and epistatic variance-covariance matrices, substituting for the traditional pedigree relationships described by Henderson to decrease prediction error of fitted mixed-models [@Henderson1975].

Consequently, the `sommer` package was used to calculate the additive genomic variance-covariance matrices and find the genomic estimated breeding value (GEBV) for each phenotype being mapped.  By converting the phenotypes to GEBV using the BLUP methodology, 
environmental effects are disentangled from the genotypic effects, increasing the power to detect QTLs in each of these traits.  To calculate BLUPs, a full model search was performed on the random terms of the relevant models defined in equation \@ref(eq:mm-year) or 
\@ref(eq:mm-year-ge) using the *Aikaike Information Criteria* (AIC) [@AIC1974].  The model with the highest AIC was subsequently used to estimate GEBV BLUPs.  In addition to deriving GEBV BLUPs from each selected mnd fitted model, random term variance estimates were 
used to calculate additive genomic heritability ($h^2$) of each trait [@DelosCampos2015].


Equation \@ref(eq:h2) displays the formula for calculating within-year narrow-sense genomic heritability, using variances estimated from fitting the mixed model \@ref(eq:mm-year).  Across year genomic heritability is calculated as shown in equation \@ref(eq:h2-ge), using 
variances estimated from the mixed model\@ref(eq:mm-year-ge).  

\begin{equation} 
    h2 = \frac{σ_g^2}{σ_g^2 + σ_ε^2}
    (\#eq:h2)
\end{equation} 

\begin{equation} 
    h2 = \frac{σ_g^2}{σ_g^2 + \frac{σ_{ge}^2}{n} + \frac{σ_ε^2}{n}}
    (\#eq:h2-ge)
\end{equation} 

## QTL mapping

QTL mapping was performed by using the `R/qtl` [CRAN](https://cran.r-project.org/) package, a software toolkit for mapping experimental crosses [@Broman2003].  A composite map consisting of SSR and SNP markers was used for QTL mapping
each respective population in order to facilitate inter-population and inter-study QTL position comparisions [@Schlautman2017].  To infer QTLs, previously modeled genotype BLUPs were substituted in lieu of 
raw phenotypes in the `R/qtl` cross table.  QTL were detected using two methods: a single-QTL interval mapping method and a model selection approach.   Both methods used Haley-Knott regression to model QTL between genetic map markers.  The single-QTL method uses
the `scanone()` function, with significant QTL determined using scanone() run against 1,000 permutations of the phenotypes in order to simulate the LOD distribution of the NULL model.
The model selection approach uses the `stepwiseqtl()` function.  This function runs a forward/backward model search algorithm by which additive and interacting terms are successively added to the model, followed by 'backward' pruning of other model terms that 
optimizes a penalized LOD score.  The penalized LOD score uses $n^{th}$ percentile thresholds derived from running scantwo() against 1,000 permutations of the phenotypes in order to control the false positive rate at $n$ percent [@Broman2009].  More complex models are penalized 
higher in order to reduce model overfitting.  For single-QTL interval mapping, the significance threshold was set for QTL with LDO scores above the $80^{th}$ permutation percentile ($p < 0.2$), while for model selection, penalized LOD scores were derived from thesholds
determined from the $95^{th}$ permutation percentile ($p < 0.05$).