---
editor_options: 
  markdown: 
    wrap: 72
---

# Results {#results}

## Correlations

```{r trait-corrplots-setup, echo=FALSE, message=FALSE, include=FALSE}
    p.cnj02.c = round(readRDS(paste0(workflow,"/traits/plots/corrplot.cnj02.rds"))$corr, digits=2)
    write_csv(tibble::as_tibble(p.cnj02.c, rownames="Trait"),"../Data/publication/tables/corrplot.cnj02.csv")
    p.cnj04.c = round(readRDS(paste0(workflow,"/traits/plots/corrplot.cnj04.rds"))$corr, digits=2)
    write_csv(tibble::as_tibble(p.cnj04.c, rownames="Trait"),"../Data/publication/tables/corrplot.cnj04.csv")
```

Prominent traits correlations in population CNJ02 highlight meaningful
phenotype relationships. The correlation heatmap of
\@ref(fig:trait-corrplots-cnj02-refined)A organizes traits into square
blocks of higher-level relationships using hierarchical agglomerative
clustering, with clusters generated by cutting the tree at height 0.6.
Only traits in non-singleton clusters are included in
\@ref(fig:trait-corrplots-cnj02-refined)A. Meaningful clusters include
the *PAC*, *TY*, and *SFY* cluster and the cluster with *MFM*, *UMFM*,
*UBM*, and *UTBM*. The barplot in figure
\@ref(fig:trait-corrplots-cnj02-refined)B shows the 15 strongest
positive and negative inter-trait correlations for CNJ02. Interesting
positive associations that emerge from the CNJ02 dataset include *SFY ×
TY*, *MFM × UBM*, *MFM × UMFM*, *UBM × UTBM*, and *MFM × UBW*. Although
not in the top 15 strongest positive correlations, *ULvW × UKLvW* shows
a moderately strong positive correlation (figure
\@ref(fig:trait-corrplots-cnj02-refined)A), evidence that the largest
berry shape categorization (a subjective trait prone to error in
recording) is consistent with that of largest berry length versus width
(an objective trait that is easy to record). Interesting negative
associations that emerge from the CNJ02 dataset include *TY × PAC*,
*UCLP × UCLS*, *SFY × PAC*, *SFY × Tacy*, *UNAB × PAC*, *MFM × PAC*,
*MFM × UNAF*, and *MFM × UBBL*.

```{r trait-corrplots-cnj0x-refined-annotate, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_annotations}
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj02.nonsingletons.squares.png"),"A",55,15))
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj02.flat.png"),"B",55,15))
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.squares.png"),"C",55,15))
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj04.flat.png"),"D",55,15))
```

```{r trait-corrplots-cnj0x-refined-montage, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_montages}
    system2("montage", 
            args=c("-background","white","-mode","concatenate", "-tile","2x2",
                   paste0(params$workflow,"/traits/plots/corrplot.cnj02.nonsingletons.squares.A.png"),
                   paste0(params$workflow,"/traits/plots/corrplot.cnj02.flat.B.png"),
                   paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.squares.C.png"),
                   paste0(params$workflow,"/traits/plots/corrplot.cnj04.flat.D.png"),
                   paste0("../Data/publication/figures/corrplot.cnj0x.nonsingletons.ABCD.png")))
```

```{r trait-corrplots-cnj0x-refined, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Trait phenotype correlation heatmaps for traits in cranberry populations CNJ02 (A and B) and CNJ04 (C and D).  Only traits found within non-singleton hierarchical clusters are shown in A and C, where black squares delineate clusters.  The 15 largest positive and 15 smallest negative correlations are shown in B and D.  Only pairwise trait correlations with a p-value less than 0.05 are displayed."}
    knitr::include_graphics("../Data/publication/figures/corrplot.cnj0x.nonsingletons.ABCD.png")
```

Noteworthy inter-trait correlations in population CNJ04 are shown in
figure \@ref(fig:trait-corrplots-cnj04-refined). The heatmap of subplot
\@ref(fig:trait-corrplots-cnj04-refined)A organizes traits into square
blocks of higher-level relationships using hierarchical agglomerative
clustering, with clusters generated by cutting the tree at height 0.6.
Only traits in non-singleton clusters are included in
\@ref(fig:trait-corrplots-cnj04-refined)A. Meaningful clusters include
the *TY*, and *SFY*, *UNB*, and *UTBM* cluster and the *MFM*, *UBM*, and
*UMFM* cluster. Compared to CNJ02, CNJ04 has weaker correlations for
traits *UBM x UBW* (`r p.cnj04.c['UBM','UBW']` vs.
`r p.cnj02.c['UBM','UBW']`) and *UBM × UBL* (`r p.cnj04.c['UBM','UBL']`
vs. `r p.cnj02.c['UBM','UBL']`), and is not in the top 15 positive
correlations in figure \@ref(fig:trait-corrplots-cnj04-refined)B.
Interesting inter-trait negative correlations include *ULvW × UKSO*,
*Brix × UBL*, *PFR × UNB*, *Brix × UNS*.

```{r trait-corrplots-cnj04-refined-annotate, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_annotations}
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.squares.png"),"A",55,15))
    system2("misc/annotatefig.sh", args=c(paste0(params$workflow,"/traits/plots/corrplot.cnj04.flat.png"),"B",55,15))
```

```{r trait-corrplots-cnj04-refined-montage, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_montages}
    system2("montage", args=c("-background","white","-mode","concatenate", "-tile","2x1",paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.squares.A.png"), paste0(params$workflow,"/traits/plots/corrplot.cnj04.flat.B.png"), paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.AB.png")))
```

```{r trait-corrplots-cnj04-refined, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="A) Trait phenotype correlation heatmaps for traits in population CNJ04 and B) Top 15 positive and negative correlations for CNJ04.  Only traits that formed non-singleton clusters in hierarchical clustering are shown in A).  Black squares emphasize traits that cluster on their correlations."}
    knitr::include_graphics(paste0(params$workflow,"/traits/plots/corrplot.cnj04.nonsingletons.AB.png"))
```

## Heritabilities

```{r h2-collate-summarize, echo=FALSE, message=FALSE, include=FALSE}
    #NOTE: The following code should be moved to QTLPipeline.Rmd
    cnj02.brcorrs <- read_csv(paste0(cnj02_workflow,"/traits/blups_raw_corrs.csv"))
    cnj02.h2 <- read_csv(paste0(cnj02_workflow,"/traits/h2.csv")) %>% 
                    dplyr::select(model,label_short,h2)
    cnj02.ay.h2 <- cnj02.h2 %>% filter(model=='all-years') %>% arrange(desc(h2)) #Sort by descending h2 for all-years model
    cnj02.combined <- read_csv(paste0(cnj02_workflow,"/traits/h2.csv")) %>% 
                        left_join(cnj02.brcorrs, by=c("model","trait")) %>%
                        dplyr::select(c(model,label_short,h2,brcorr)) %>%
                        pivot_longer(c(h2,brcorr), names_to="type", values_to="value") %>%
                        mutate(population="cnj02",
                               trait=factor(label_short, levels=rev(cnj02.ay.h2$label_short), ordered=TRUE),
                               type=factor(type, levels=c("brcorr","h2"), labels=c("Spearman ρ","Heritability")))
    cnj04.brcorrs <- read_csv(paste0(cnj04_workflow,"/traits/blups_raw_corrs.csv"))
    cnj04.h2 <- read_csv(paste0(cnj04_workflow,"/traits/h2.csv")) %>% 
                    dplyr::select(model,label_short,h2)
    cnj04.ay.h2 <- cnj04.h2 %>% filter(model=='all-years') %>% arrange(desc(h2)) #Sort by descending h2 for all-years model
    cnj04.combined <- read_csv(paste0(cnj04_workflow,"/traits/h2.csv")) %>% 
                        left_join(cnj04.brcorrs, by=c("model","trait")) %>%
                        dplyr::select(c(model,label_short,h2,brcorr)) %>%
                        pivot_longer(c(h2,brcorr), names_to="type", values_to="value") %>%
                        mutate(population="cnj04",
                               trait=factor(label_short, levels=rev(cnj04.ay.h2$label_short), ordered=TRUE),
                               type=factor(type, levels=c("brcorr","h2"), labels=c("Spearman ρ","Heritability")))
    cnj0x.h2.sum.nay <- rbind(cnj02.h2,cnj04.h2) %>%
                            filter(model != 'all-years') %>%
                            group_by(label_short) %>%
                            summarize(h2.min=min(h2),h2.max=max(h2),h2.mean=mean(h2))
    cnj0x.h2.sum.ay <- rbind(cnj02.h2,cnj04.h2) %>%
                            filter(model == 'all-years') %>%
                            group_by(label_short) %>%
                            summarize(h2.min=min(h2),h2.max=max(h2),h2.mean=mean(h2))
    cnj0x.h2.sum.nayay <- cnj0x.h2.sum.nay %>%
                            inner_join(cnj0x.h2.sum.ay, by="label_short", suffix=c(".nay",".ay"))
    cnj0x.h2.sum <- rbind(cnj02.h2,cnj04.h2) %>%
                    group_by(model, label_short) %>%
                    summarize(h2.min=min(h2),h2.max=max(h2),h2.mean=mean(h2))
    cnj0x.h2 <- cnj02.h2 %>%
                    inner_join(cnj04.h2, by=c("model","label_short"), suffix=c(".2",".4")) %>%
                    inner_join(cnj0x.h2.sum, by=c("model","label_short")) %>%
                    arrange(desc(h2.min))
    cnj0x.h2.all <- cnj0x.h2 %>%
                        filter(model=="all-years") %>%
                        mutate(h2.min=round(h2.min,digits=2)) %>%
                        mutate(trait=factor(label_short, levels=rev(label_short), ordered=TRUE))
    #The following are for inserting inline results in the text of the document
    cnj04.combined.ul.h2 <- round((cnj04.ay.h2 %>% filter(label_short == "UL"))$h2, digits=2)
    cnj04.combined.ul.brcorrs <- round((cnj04.brcorrs %>% filter(model == 'all-years') %>% filter(trait == "upright_length"))$brcorr, digits=2)

    load(paste0(cnj02_workflow,"/.RData.01_genBLUP"))
    model.collated.long.tb.2 <- model.collated.long.tb %>% mutate(population="cnj02")
    model.collated.long.stats.tb.2 <- model.collated.long.stats.tb %>% mutate(population="cnj02")
    model.collated.long.ylims.tb.2 <- model.collated.long.ylims.tb %>% mutate(population="cnj02")
    load(paste0(cnj04_workflow,"/.RData.01_genBLUP"))
    model.collated.long.tb.4 <- model.collated.long.tb %>% mutate(population="cnj04")
    model.collated.long.stats.tb.4 <- model.collated.long.stats.tb %>% mutate(population="cnj04")
    model.collated.long.ylims.tb.4 <- model.collated.long.ylims.tb %>% mutate(population="cnj04")

    model.collated.long.tb.both <- bind_rows(model.collated.long.tb.2,model.collated.long.tb.4) %>%
                                        filter(model == "all-years") %>% 
                                        filter(h2 != 0)
    model.collated.long.stats.tb.both <- bind_rows(model.collated.long.stats.tb.2,model.collated.long.stats.tb.4)
    model.collated.long.ylims.tb.both <- bind_rows(model.collated.long.ylims.tb.2,model.collated.long.ylims.tb.4)

    cnj0x.h2.top.all <- cnj0x.h2.all %>% 
                            filter(h2.min > median(h2.mean.ay)) %>%
                            mutate(trait = factor(trait, levels=rev(trait)))
    cnj0x.median.h2.ay <- median(cnj0x.h2.sum.nayay$h2.mean.ay)
    cnj0x.median.h2.ay.r <- round(cnj0x.median.h2.ay,2)
    cnj0x.h2.top.nayay <- cnj0x.h2.sum.nayay %>%  #nay means 'not all-years' and represents the average of all models except 'all-years', while ay means 'all-years'
                            filter(h2.mean.ay >= cnj0x.median.h2.ay) %>% 
                            arrange(desc(h2.mean.ay)) %>% 
                            select(label_short,h2.mean.ay,h2.mean.nay) %>%
                            mutate(trait=factor(label_short, levels=rev(label_short), ordered=TRUE))
    trait2ShortMap = levels(model.collated.long.tb$label_short)
    names(trait2ShortMap) = levels(model.collated.long.tb$trait)

    #Reset workflow as it is overwritten
    workflow = params$workflow
```

```{r echo=FALSE, message=FALSE, include=TRUE}
    load(paste0(cnj02_workflow,"/.RData.01_genBLUP"))
    gs_all <- model.collated.long.tb %>% 
        filter(model == 'all-years') %>%
        group_by(model,trait,type) %>%
        filter(min(value) != max(value)) %>%
        ungroup() %>%
        group_nest(trait) %>%
        left_join(model.collated.long.stats.tb, by="trait") %>% 
        left_join(model.collated.long.ylims.tb, by="trait") %>%
        group_by(trait) %>%
        do(
            plot = ggplot(.$data[[1]], aes(x=factor(type), y=value, fill=factor(type))) +
                geom_boxplot(alpha=0.4, position=position_dodge(1)) +
                geom_jitter(alpha=0.1, position=position_jitterdodge(jitter.width=0.5, dodge.width=1)) +
                geom_text(data=.$stats[[1]]%>%filter(model=='all-years'), position=position_dodge(1), aes(y=min, label=min_geno), size=8, fontface='bold', vjust=1.5, hjust=1.5, angle=45) +
                geom_text(data=.$stats[[1]]%>%filter(model=='all-years'), position=position_dodge(1), aes(y=max, label=max_geno), size=8, fontface='bold', vjust=-1.5, hjust=0, angle=45) +
                geom_blank(data=.$ylims[[1]] %>% filter((limit == "min") | (limit == "max"))) +
                ylab("") +
                xlab("") + 
                guides(fill='none') +
                ggtitle(label=.$data[[1]]$label_short) +
                theme(  axis.text.x = element_text(face="bold", size=32, angle = 60, hjust = 1),
                        axis.text.y = element_text(face="bold", size=32),
                        legend.title = element_text(fac="bold", size=32),
                        legend.text = element_text(fac="bold", size=24),
                        legend.key.size = ggplot2::unit(5,"cm"),
                        axis.title  = element_text(face="bold",size=48),
                        plot.title   = element_text(face="bold",size=52, hjust=0.5),
                        plot.subtitle = element_text(size=48, hjust = 0.5),
                        plot.margin = ggplot2::unit(c(1,1,1,1),"cm")))
    gs_all <- gs_all %>% 
                mutate(label_short = factor(trait2ShortMap[as.character(trait)], levels=rev(levels(cnj0x.h2.top.all$trait)), ordered=TRUE)) %>% 
                filter(label_short %in% cnj0x.h2.top.all[1:8,]$label_short) %>% 
                arrange(label_short)
    pg <- plot_grid(plotlist=gs_all$plot,nrow=2,ncol=4,rel_widths=c(1,1,1,1))
    png(filename=paste0(cnj02_workflow,'/traits/plots/blups_collated.boxplot.h2.top.png'), width=3840, height=2560, bg="white")
    pg
    dev.off()
    #CNJ04 Population
    load(paste0(cnj04_workflow,"/.RData.01_genBLUP"))
    gs_all <- model.collated.long.tb %>% 
        filter(model == 'all-years') %>%
        group_by(model,trait,type) %>%
        filter(min(value) != max(value)) %>%
        ungroup() %>%
        group_nest(trait) %>%
        left_join(model.collated.long.stats.tb, by="trait") %>% 
        left_join(model.collated.long.ylims.tb, by="trait") %>%
        group_by(trait) %>%
        do(
            plot = ggplot(.$data[[1]], aes(x=factor(type), y=value, fill=factor(type))) +
                geom_boxplot(alpha=0.4, position=position_dodge(1)) +
                geom_jitter(alpha=0.1, position=position_jitterdodge(jitter.width=0.5, dodge.width=1)) +
                geom_text(data=.$stats[[1]]%>%filter(model=='all-years'), position=position_dodge(1), aes(y=min, label=min_geno), size=8, fontface='bold', vjust=1.5, hjust=1.5, angle=45) +
                geom_text(data=.$stats[[1]]%>%filter(model=='all-years'), position=position_dodge(1), aes(y=max, label=max_geno), size=8, fontface='bold', vjust=-1.5, hjust=0, angle=45) +
                geom_blank(data=.$ylims[[1]] %>% filter((limit == "min") | (limit == "max"))) +
                ylab("") +
                xlab("") + 
                guides(fill='none') +
                ggtitle(label=.$data[[1]]$label_short) +
                theme(  axis.text.x = element_text(face="bold", size=32, angle = 60, hjust = 1),
                        axis.text.y = element_text(face="bold", size=32),
                        legend.title = element_text(fac="bold", size=32),
                        legend.text = element_text(fac="bold", size=24),
                        legend.key.size = ggplot2::unit(5,"cm"),
                        axis.title  = element_text(face="bold",size=48),
                        plot.title   = element_text(face="bold",size=52, hjust=0.5),
                        plot.subtitle = element_text(size=48, hjust = 0.5),
                        plot.margin = ggplot2::unit(c(1,1,1,1),"cm")))
    gs_all <- gs_all %>% 
                mutate(label_short = factor(trait2ShortMap[as.character(trait)], levels=rev(levels(cnj0x.h2.top.all$trait)), ordered=TRUE)) %>% 
                filter(label_short %in% cnj0x.h2.top.all[1:8,]$label_short) %>% 
                arrange(label_short)
    pg <- plot_grid(plotlist=gs_all$plot,nrow=2,ncol=4,rel_widths=c(1,1,1,1))
    png(filename=paste0(cnj04_workflow,'/traits/plots/blups_collated.boxplot.h2.top.png'), width=3840, height=2560, bg="white")
    pg
    dev.off()
    workflow <- params$workflow
```

For both populations, the highest heritabilities were found for the
upright yield-related traits and the plot traits (figure
\@ref(fig:trait-h2-cnj0x)). Generally, the *all year* models (equation
\@ref(eq:mm-year-ge)) had the highest heritabilities when compared to
the fitted models for individual years. Traits with consistently high
heritability in both populations are displayed in figure
\@ref(fig:h2-top) next to boxplots of their associated phenotypes and
BLUPs. The heritabilities are sorted from highest to lowest and
represent the minimum heritability between the two populations. The
BLUPs and phenotype boxplot distributions are standardized against the
raw phenotype trait values in order to demonstrate that models with
higher heritabilities encapsulate more of the phenotypic variation.
Trait *ULvW* has an extraordinarily high heritability in both
populations, followed by *UBL* and *UNS*. *Tacy*, a desirable trait
associated with fruit color, also shows evidence of high heritability.
High heritabilities in *UBM*, *UMFM*, and *TY* indicate strong potential
fordplyr::selecting for higher yields in both of these two populations.
Although only modest, the heritability of traits like *SFY* and *PFR*
show some potential fordplyr::selecting rot-resistant varieties for growing in
regions where berry rot is a problem. Fruit quality traits such as *TA*
and *Brix* show very low *all year* heritabilities in population CNJ02
but higher heritability in CNJ04.

Figure \@ref(fig:h2-brcorr) displays *all year* model heritabilities
sorted highest to lowest for each population separately next to the
calculated Spearman rank correlation coefficient for the genotype BLUPs
relative to their raw phenotypes. In spite of low heritability, traits
such as *UL* in CNJ04 continue to show moderate to strong correlation
between the BLUPs and their raw trait values (*e.g.*, \$h\^2 = \$
`r cnj04.combined.ul.h2`, \$ρ = \$ `r cnj04.combined.ul.brcorrs`),
indicating that the models continue to maintain good genotype stability
when presented with traits that have incredibly low heritability and
that issues with genotype × environment interaction are minimal.

```{r h2-top-pre, echo=FALSE, message=FALSE, include=TRUE}
    p1 <- ggplot(cnj0x.h2.top.nayay, aes(x=h2.mean.ay,y=trait)) + 
            geom_bar(aes(fill=trait),stat="identity", alpha=0.75, show.legend=FALSE) +
            geom_text(aes(label=round(h2.mean.ay,2),fontface="bold"), nudge_x=0.06) +
            scale_y_discrete(position = "right") +
            scale_x_continuous(breaks=c(0,0.25,0.5,0.75,1.0)) +
            xlim(0,1.05) +
            theme_minimal() +
            theme(axis.title = element_text(face="bold",size=14),
                axis.text.y  = element_text(face="bold",size=16,angle=0),
                axis.text.x  = element_text(size=12),
                plot.margin = margin(r=0, unit="in")) +
            labs(x="'All Years' Model Heritabilities",
                y="")
    p2 <- ggplot(cnj0x.h2.top.nayay, aes(x=h2.mean.nay,y=trait)) + 
            geom_bar(aes(fill=trait),stat="identity", alpha=0.75, show.legend=FALSE) +
            geom_text(aes(label=round(h2.mean.nay,2),fontface="bold"), nudge_x=-0.06) +
            scale_y_discrete(position = "right") +
            scale_x_continuous(breaks=c(0,0.25,0.5,0.75,1.0)) +
            scale_x_reverse() +
            xlim(1.05,0) +
            theme_minimal() +
            theme(axis.title = element_text(face="bold",size=14),
                axis.text.y  = element_blank(),
                axis.text.x  = element_text(size=12),
                plot.margin = margin(l=0.25, unit="in")) +
            labs(x="Mean Heritabilities across Year Models",
                y="")

    ggsave(filename="../Data/publication/figures/cnj0x.h2.top.ay.png", plot=p1, device="png", bg="white", width=7, height=5, units="in", dpi=300)
    ggsave(filename="../Data/publication/figures/cnj0x.h2.top.nay.png", plot=p2, device="png", bg="white", width=6.4, height=5, units="in", dpi=300)
```

```{r h2-top, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Mean heritability between cranberry populations CNJ02 and CNJ04 A. of ‘all years’ model and B. across within-year models.  Only the top 50% of traits based on mean within-year model heritability are displayed."}
    system2("misc/annotatefig.sh", args=c("../Data/publication/figures/cnj0x.h2.top.ay.png","A",75,15,"nb"))
    system2("misc/annotatefig.sh", args=c("../Data/publication/figures/cnj0x.h2.top.nay.png","B",75,15,"nb"))
    system2("montage", args=c("-background","white","-mode","concatenate", "-tile","2x1","../Data/publication/figures/cnj0x.h2.top.ay.A.png","../Data/publication/figures/cnj0x.h2.top.nay.B.png","../Data/publication/figures/cnj0x.h2.top.nayay.AB.png"))
    knitr::include_graphics("../Data/publication/figures/cnj0x.h2.top.nayay.AB.png")
```

```{r h2-brcorr-plots, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
    #NOTE: The following code should be moved to QTLPipeline.Rmd
    p2 <- ggplot(cnj02.combined %>% filter(model == "all-years"), aes(x=value, y=trait, group=factor(type), fill=type)) + 
        geom_bar(stat="identity", position=position_dodge(), alpha=0.75) +
        theme_minimal() +
        theme(plot.title = element_text(face="bold",size=14,hjust=0.5),
              axis.title = element_text(face="bold",size=12),
              legend.title = element_text(face="bold",size=12),
              axis.text.y  = element_text(size=12,angle=25),
              axis.text.x  = element_text(size=12,angle=60)) +
        labs(title="CNJ02",
             x="",
             y="Traits",
             fill="Type")
    ggsave(filename="../Data/publication/figures/cnj02.barplot.h2spearman.png", plot=p2, device="png", bg="white", width=12, height=20, units="cm", dpi=300)

    p4 <- ggplot(cnj04.combined %>% filter(model == "all-years"), aes(x=value, y=trait, group=factor(type), fill=type)) + 
        geom_bar(stat="identity", position=position_dodge(), alpha=0.75) +
        theme_minimal() +
        theme(plot.title = element_text(face="bold",size=14,hjust=0.5),
              axis.title = element_text(face="bold",size=12),
              legend.title = element_text(face="bold",size=12),
              axis.text.y  = element_text(size=12,angle=25),
              axis.text.x  = element_text(size=12,angle=60)) +
        labs(title="CNJ04",
             x="",
             y="Traits",
             fill="Type")
    ggsave(filename="../Data/publication/figures/cnj04.barplot.h2spearman.png", plot=p4, device="png", bg="white", width=12, height=20, units="cm", dpi=300)
```

```{r h2-brcorr-annotate, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
    system2("misc/annotatefig.sh", args=c("../Data/publication/figures/cnj02.barplot.h2spearman.png","A"))
    system2("misc/annotatefig.sh", args=c("../Data/publication/figures/cnj04.barplot.h2spearman.png","B"))
```

```{r h2-brcorr-montage, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE, eval=FALSE}
    system2("montage", args=c("-background","white","-geometry","1421x2366+20+0","-tile","2x1","../Data/publication/figures/cnj02.barplot.h2spearman.A.png","../Data/publication/figures/cnj04.barplot.h2spearman.B.png","../Data/publication/figures/cnj0x.barplot.h2spearman.png"))
```

```{r h2-brcorr, echo=FALSE, message=FALSE, warning=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Heritabilities and Spearman rank correlation coefficients ($ρ$) for traits in populations A) CNJ02 and B) CNJ04.  Correlation coefficients represent the associations between genotype BLUPs and their associated phenotypes for the *all years* mixed model.", eval=FALSE}
    knitr::include_graphics("../Data/publication/figures/cnj0x.barplot.h2spearman.png")
```

```{r trait-h2-cnj0x-yield-def, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    if( params$h2_plots_polar == TRUE ) {
        imageA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.polar.upright_yield.CNJ02.png")
        imageAA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.polar.upright_yield.CNJ02.A.png")
        imageB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.polar.upright_yield.CNJ04.png")
        imageBB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.polar.upright_yield.CNJ04.B.png")
        imageCombined <- "../Data/publication/figures/h2_barplot.polar.upright_yield.CNJ0x.AB.png"
    } else {
        imageA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.cartesian.upright_yield.CNJ02.png")
        imageAA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.cartesian.upright_yield.CNJ02.A.png")
        imageB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.cartesian.upright_yield.CNJ04.png")
        imageBB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.cartesian.upright_yield.CNJ04.B.png")
        imageCombined <- "../Data/publication/figures/h2_barplot.cartesian.upright_yield.CNJ0x.AB.png"
    }

```

```{r trait-h2-cnj0x-yield-annotate, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE}
    getwd()
    system2("misc/annotatefig.sh", args=c(imageA,"A",75,25,"nb"))
    system2("misc/annotatefig.sh", args=c(imageB,"B",75,25,"nb"))
```

```{r trait-h2-cnj0x-yield-montage, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    system2("montage", args=c("-background","white","-geometry","1500x2700+20+20","-tile","2x1",imageAA,imageBB,imageCombined))
```
    cnj02_h2_ay_spearman <- cnj02_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_h2_ay_pearson <- cnj02_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")

```{r trait-h2-cnj0x-yield, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Heritabilities for upright-related yield traits (A and C) and plot-sampled traits (B and D) in populations CNJ02 (A and B) and CNJ04 (C and D).", eval=FALSE}
    knitr::include_graphics(imageCombined)
```

```{r trait-h2-cnj0x-plot-def, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    if( params$h2_plots_polar == TRUE ) {
        imageA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.polar.plot_traits.CNJ02.png")
        imageAA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.polar.plot_traits.CNJ02.A.png")
        imageB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.polar.plot_traits.CNJ04.png")
        imageBB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.polar.plot_traits.CNJ04.B.png")
        imageCombined <- "../Data/publication/figures/h2_barplot.polar.plot_traits.CNJ0x.AB.png"
    } else {
        imageA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.cartesian.plot_traits.CNJ02.png")
        imageAA <- paste0(cnj02_workflow,"/traits/plots/h2_barplot.cartesian.plot_traits.CNJ02.A.png")
        imageB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.cartesian.plot_traits.CNJ04.png")
        imageBB <- paste0(cnj04_workflow,"/traits/plots/h2_barplot.cartesian.plot_traits.CNJ04.B.png")
        imageCombined <- "../Data/publication/figures/h2_barplot.cartesian.plot_traits.CNJ0x.AB.png"
    }

```

```{r trait-h2-cnj0x-plot-annotate, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE}
    system2("misc/annotatefig.sh", args=c(imageA,"A",75,25,"nb"))
    system2("misc/annotatefig.sh", args=c(imageB,"B",75,25,"nb"))
```

```{r trait-h2-cnj0x-plot-montage, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    system2("montage", args=c("-background","white","-geometry","1354x2704+20+20","-tile","2x1",imageAA,imageBB,imageCombined))
```

```{r trait-h2-cnj0x-plot, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Heritabilities for upright-related yield traits (A and C) and plot-sampled traits (B and D) in populations CNJ02 (A and B) and CNJ04 (C and D).", eval=FALSE}
    knitr::include_graphics(imageCombined)
```

```{r h2-compare, echo=FALSE, include=FALSE, eval=FALSE}
    cnj02_h2_a = read_csv("../Workflows/9/traits/h2.csv", show_col_types=F)
    cnj02_h2_ay_a = cnj02_h2_a %>% filter(model == "all-years")
    cnj02_h2_ny_a = cnj02_h2_a %>% filter(model != "all-years")
    cnj02_h2_b = read_csv("../Workflows/11/traits/h2.csv", show_col_types=F)
    cnj02_h2_ay_b = cnj02_h2_b %>% filter(model == "all-years")
    cnj02_h2_ny_b = cnj02_h2_b %>% filter(model != "all-years")
    cnj02_h2_join = cnj02_h2_a %>% full_join(cnj02_h2_b, by=c("trait","model"))
    cnj02_h2_ay_join = cnj02_h2_ay_a %>% full_join(cnj02_h2_ay_b, by=c("trait","model"))
    cnj02_h2_ny_join = cnj02_h2_ny_a %>% full_join(cnj02_h2_ny_b, by=c("trait","model"))
    cnj02_h2_spearman <- cnj02_h2_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_h2_pearson <- cnj02_h2_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj02_h2_ay_spearman <- cnj02_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_h2_ay_pearson <- cnj02_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj02_h2_ny_spearman <- cnj02_h2_ny_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_h2_ny_pearson <- cnj02_h2_ny_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    
    cnj04_h2_a = read_csv("../Workflows/10/traits/h2.csv", show_col_types=F)
    cnj04_h2_ay_a = cnj04_h2_a %>% filter(model == "all-years")
    cnj04_h2_ny_a = cnj04_h2_a %>% filter(model != "all-years")
    cnj04_h2_b = read_csv("../Workflows/12/traits/h2.csv", show_col_types=F)
    cnj04_h2_ay_b = cnj04_h2_b %>% filter(model == "all-years")
    cnj04_h2_ny_b = cnj04_h2_b %>% filter(model != "all-years")
    cnj04_h2_join = cnj04_h2_a %>% full_join(cnj04_h2_b, by=c("trait","model"))
    cnj04_h2_ay_join = cnj04_h2_ay_a %>% full_join(cnj04_h2_ay_b, by=c("trait","model"))
    cnj04_h2_ny_join = cnj04_h2_ny_a %>% full_join(cnj04_h2_ny_b, by=c("trait","model"))
    cnj04_h2_spearman <- cnj04_h2_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_h2_pearson <- cnj04_h2_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj04_h2_ay_spearman <- cnj04_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_h2_ay_pearson <- cnj04_h2_ay_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj04_h2_ny_spearman <- cnj04_h2_ny_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_h2_ny_pearson <- cnj04_h2_ny_join %>% dplyr::select(h2.x,h2.y) %>% cor(use="pairwise.complete.obs",method="pearson")
```
##Breeding Value Estimates

Population CNJ02 and CNJ04 both displayed evidence of transgressive
segregation in many of the traits. Tables
\@ref(tab:blups-table-cnj02-all-top) and
\@ref(tab:blups-table-cnj04-all-top) show summary statistics for traits
and their associated BLUPs for the *all year* model. Rows are sorted by
decreasing heritability, and only models with significant genotypic
effects (*p \< 0.05*) and heritability above `r cnj0x.median.h2.ay.r` are displayed. Optimal
model random effect terms used in fitting each model are shown. Raw
trait associated statistics are subscripted with the letter *r*, and
BLUP associated statistics are subscripted with the letter *b* and are
italicized. The top and bottom five representative genotypes, both for
raw traits and for genotype BLUPs, are also shown.

### Summary Statistics

```{r blups-table-cnj02-all-top, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="75%"}
clabel <- knitr::opts_current$get('label')
caption <- paste0("CNJ02 Population Trait and BLUP Summary for *all years* model with heritability above ",cnj0x.median.h2.ay.r,". Only models with significant (*p < 0.05*) genotypic effects are displayed.")
load(file=paste0(cnj02_workflow,"/.RData.01_genBLUP"))
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
tbl <- model.collated.summary.wide.tb %>% 
        filter(model == 'all-years' &
               label_short %in% cnj0x.h2.top.nayay$trait &
               h2 > cnj0x.median.h2.ay) %>%
        mutate(label=factor(label_short, levels=levels(cnj0x.h2.top.nayay$trait), ordered=TRUE)) %>%
        arrange(desc(as.integer(label))) %>%
        filter(GLRpvalue < 0.05)
scipen.old <- getOption('scipen')
options(scipen=-2)
ftbl <- generateReducedTable(tbl, caption=caption,  tfont_size=table.font_size)
if(!is.null(getOption("knitr.in.progress"))) {
    print_table(ftbl, clabel, caption)
} else {
    if(is_word_output()) {
        save_as_docx(ftbl,path="../Data/publication/tables/blups-table-cnj02-all-top.docx")
    }
}
options(scipen=scipen.old)
```

```{r blups-table-cnj04-all-top, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="75%"}
clabel <- knitr::opts_current$get('label')
caption <- paste0("CNJ04 Population Trait and BLUP Summary for *all years* model with heritability above ",cnj0x.median.h2.ay.r,".  Only models with significant ($p < 0.05$) genotypic effects are displayed.")
load(file=paste0(cnj04_workflow,"/.RData.01_genBLUP"))
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
tbl <- model.collated.summary.wide.tb %>% 
        filter(model == 'all-years' &
               label_short %in% cnj0x.h2.top.nayay$trait &
               h2 > cnj0x.median.h2.ay) %>%
        mutate(label=factor(label_short, levels=levels(cnj0x.h2.top.nayay$trait), ordered=TRUE)) %>%
        arrange(desc(as.integer(label))) %>%
        filter(GLRpvalue < 0.05)
scipen.old <- getOption('scipen')
options(scipen=-2)
ftbl <- generateReducedTable(tbl, caption=caption,  tfont_size=table.font_size)
if(!is.null(getOption("knitr.in.progress"))) {
    print_table(ftbl, clabel, caption)
} else {
    if(is_word_output()) {
        save_as_docx(ftbl,path="../Data/publication/tables/blups-table-cnj04-all-top.docx")
    }
}
options(scipen=scipen.old)
```

### Genotype Rankings

```{r blups-plot-cnj0x-rank-sum-defs, echo=FALSE, message=FALSE, include=TRUE}
rank_blups <- function(blups, traits, direction=1, prop=.10) {
    o1 <- blups %>%
                filter(!(id %in% c(P1_Name,P2_Name))) %>%
                filter(label_short %in% traits) %>%
                filter(type == "BLUPs") %>%
                mutate(id=gsub(paste0(POP_Name,"_([0-9]+_[0-9]+)"),"g\\1",id)) %>%
                group_by(model,label_short)
    if( direction >= 0 ) {
        o2 <- o1 %>% slice_max(value,prop=prop)
    } else {
        o2 <- o1 %>% slice_min(value,prop=prop)
    }
    o3 <- o2 %>%
             ungroup() %>%
             group_by(model,label_short)
    if( direction >= 0 ) {
        o4 <- o3 %>% arrange(value, .by_group=TRUE)
    } else {
        o4 <- o3 %>% arrange(desc(value), .by_group=TRUE)
    }
    o5 <- o4 %>% 

                     mutate(frank=1:length(id)) %>%
             ungroup()
    return(o5)
}

sum_ranks <- function(blups.ranked) {
    blups.sums <- blups.ranked %>% 
                 dplyr::select(model,label_short,id,frank) %>% 
                  group_by(model,id) %>% 
                  summarize(rankcount=sum(frank)) %>%
                  arrange(desc(rankcount)) %>%
                  filter(model=="all-years") %>% 
                  mutate(id=factor(id, levels=id, ordered=TRUE))
    return(blups.sums)
}

plot_rank_sums <- function(blups.ranked, blups.sums, filename) {
    p <- ggplot(blups.ranked %>% filter(model == "all-years"), aes(x=frank, y=factor(id, levels=rev(blups.sums$id), ordered=TRUE), fill=label_short)) + 
            geom_bar(stat="identity", position=position_stack(), alpha=0.9) + 
            scale_fill_brewer(
                type="qual",
                palette="Paired") +
            theme_minimal() +
            theme(plot.title = element_text(face="bold",size=10,hjust=0.5),
                axis.title = element_text(face="bold",size=9,hjust=0.5),
                axis.text.x  = element_text(size=7),
                axis.text.y  = element_text(size=7),
                legend.text=element_text(size=8),
                legend.title=element_text(face="bold",size=10)) +
            labs(x="Rank Count",
                y="Genotype",
                fill="Trait",
                title=paste0(POP_Name," Rank Sum of Top 10 Percent Genotypes"))
    ggsave(filename=filename, plot=p, device="png", bg="white", width=12, height=20, units="cm", dpi=300)
}

imageA <- "../Data/publication/figures/cnj02.blups.rank_sums.png"
imageAA <- "../Data/publication/figures/cnj02.blups.rank_sums.A.png"
imageB <- "../Data/publication/figures/cnj04.blups.rank_sums.png"
imageBB <- "../Data/publication/figures/cnj04.blups.rank_sums.B.png"
imageCombined <- "../Data/publication/figures/cnj0x.blups.rank_sums.png"
```

```{r blups-plot-cnj02-rank-sum-plot, echo=FALSE, message=FALSE, include=TRUE}
load(file=paste0(cnj02_workflow,"/.RData.01_genBLUP"))
#Traits where higher values are preferred
positive_traits <- c("UBM","SFY","Tacy","MFM","UMFM")
#Traits where lower values are preferred
negative_traits <- c("ULvW","UKEC","PFR")

pts <- rank_blups(model.collated.long.tb, positive_traits)
nts <- rank_blups(model.collated.long.tb, negative_traits, direction=-1)
pnts <- bind_rows(pts,nts)
brs <- sum_ranks(pnts)
plot_rank_sums(pnts, brs, filename=imageA)
```

```{r blups-plot-cnj04-rank-sum-plot, echo=FALSE, message=FALSE, include=TRUE}
load(file=paste0(cnj04_workflow,"/.RData.01_genBLUP"))
#Traits where higher values are preferred
positive_traits <- c("UBM","SFY","Tacy","MFM","UMFM","Brix")
#Traits where lower values are preferred
negative_traits <- c("ULvW","UKEC","PFR")

pts <- rank_blups(model.collated.long.tb, positive_traits)
nts <- rank_blups(model.collated.long.tb, negative_traits, direction=-1)
pnts <- bind_rows(pts,nts)
brs <- sum_ranks(pnts)
plot_rank_sums(pnts, brs, filename="../Data/publication/figures/cnj04.blups.rank_sums.png")
workflow = params$workflow #Reset
```

```{r blups-plot-cnj0x-rank-sum-plot-annotate, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_annotations}
    system2("misc/annotatefig.sh", args=c(imageA,"A"))
    system2("misc/annotatefig.sh", args=c(imageB,"B"))
```

```{r blups-plot-cnj0x-rank-sum-plot-montage, echo=FALSE, message=FALSE, include=TRUE, eval=params$fig_gen_montage}
    system2("montage", args=c("-background","white","-geometry","1425x2370+20+20","-tile","2x1",imageAA,imageBB,imageCombined))
```

A distribution of the top genotypes broken down by ranking the top BLUPs
indicates a few strong candidates fordplyr::selection in a breeding program
\@ref(fig:blups-plot-cnj0x-rank-sum-plot). For population CNJ02,
genotypes *CNJ02_1\_38*, *CNJ02_1\_82*, *CNJ02_1\_17*, *CNJ02_1\_77*,\
and *CNJ02_1\_57* have high combined BLUP ranking for largest berry
traits such as *UBM* and *UMFM*, in addition to the mean fruit mass of
the plot-sampled traits (*MFM*). When considering other traits important
to breeding, such as larger overall sound fruit yield (*SFY*) and low
fruit rot (*PFR*), genotype *CNJ02_1\_118* shows promise. Genotypes
*CNJ02_1\_138* and *CNJ02_1\_100* are ideal nominees when considering
berry quality characteristics such as *Tacy* and rounder berries (low
*UKEC* and low *ULvW*). Because the CNJ04 population is significantly
smaller, less candidates are available with stacked traits. However,
*CNJ04_21_10* and *CNJ04_2\_5* shows excellent potential asdplyr::selection
candidates for berries with high *MFM* and largest berry mass (*UBM*),
while *CNJ04_21_10* has high ranking with A *SFY*. *CNJ04_21_32*
combines high *Tacy* with high *MFM* and rounder berries (low *ULvW*).
*CNJ04_2\_24* combines high *Tacy* with low fruit rot (*PFR*), along
with rounder berries (low *UKEC* and low *ULvW*). *CNJ04_2\_12* stacks
higher *Brix* with rounder berries (low *UKEC*).

```{r blups-plot-cnj0x-rank-sum-plot-defs, echo=FALSE, message=FALSE, include=TRUE}
caption="Rank-sum of top 10 percent of genotypes for adplyr::selected subset of trait BLUPs for the *all year* model.  A) Population CNJ02 and b) Population CNJ04.  Genotype identifier is abbreviated: g<num> => CNJ02_<num>.  eg: g1_77 => CNJ02_1_77"
if( is_pdf_output() ) {
    #If we don't properly escape the captions, then the latex code that is generated isn't correct -- it inserts a
    #\texbackslash in front of everything
    caption = knitr:::escape_latex(caption)
}
```

```{r blups-plot-cnj0x-rank-sum-plot, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap=caption}
    knitr::include_graphics(imageCombined)
```

## 
```{r blups-compare, echo=FALSE, include=FALSE, eval=F}
    cnj02_blups_a = readRDS("../Workflows/9/traits/blups_collated.wide.rds") %>% 
                        dplyr::filter(GLRpvalue < 0.05) %>%
                        dplyr::select(genotype,model,trait,blup)
    cnj02_blups_b = readRDS("../Workflows/11/traits/blups_collated.wide.rds") %>%
                        dplyr::filter(GLRpvalue < 0.05) %>%
                        dplyr::select(genotype,model,trait,blup)
    cnj02_blups_ny_a =  cnj02_blups_a %>% dplyr::filter(model != "all-years")
    cnj02_blups_ay_a =  cnj02_blups_a %>% dplyr::filter(model == "all-years")
    cnj02_blups_ny_b =  cnj02_blups_b %>% dplyr::filter(model != "all-years")
    cnj02_blups_ay_b =  cnj02_blups_b %>% dplyr::filter(model == "all-years")
    cnj02_blups_join = cnj02_blups_a %>% full_join(cnj02_blups_b, by=c("genotype","trait","model"))
    cnj02_blups_ay_join = cnj02_blups_ay_a %>% full_join(cnj02_blups_ay_b, by=c("genotype","trait","model"))
    cnj02_blups_ny_join = cnj02_blups_ny_a %>% full_join(cnj02_blups_ny_b, by=c("genotype","trait","model"))
    cnj02_blups_spearman <- cnj02_blups_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_blups_pearson <- cnj02_blups_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj02_blups_ay_spearman <- cnj02_blups_ay_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_blups_ay_pearson <- cnj02_blups_ay_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj02_blups_ny_spearman <- cnj02_blups_ny_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj02_blups_ny_pearson <- cnj02_blups_ny_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    
    cnj04_blups_a = readRDS("../Workflows/10/traits/blups_collated.wide.rds") %>% 
                        dplyr::filter(GLRpvalue < 0.05) %>%
                        dplyr::select(genotype,model,trait,blup)
    cnj04_blups_b = readRDS("../Workflows/12/traits/blups_collated.wide.rds") %>% 
                        dplyr::filter(GLRpvalue < 0.05) %>%
                        dplyr::select(genotype,model,trait,blup)
    cnj04_blups_ny_a =  cnj04_blups_a %>% dplyr::filter(model != "all-years")
    cnj04_blups_ay_a =  cnj04_blups_a %>% dplyr::filter(model == "all-years")
    cnj04_blups_ny_b =  cnj04_blups_b %>% dplyr::filter(model != "all-years")
    cnj04_blups_ay_b =  cnj04_blups_b %>% dplyr::filter(model == "all-years")
    cnj04_blups_join = cnj04_blups_a %>% full_join(cnj04_blups_b, by=c("genotype","trait","model"))
    cnj04_blups_ay_join = cnj04_blups_ay_a %>% full_join(cnj04_blups_ay_b, by=c("genotype","trait","model"))
    cnj04_blups_ny_join = cnj04_blups_ny_a %>% full_join(cnj04_blups_ny_b, by=c("genotype","trait","model"))
    cnj04_blups_spearman <- cnj04_blups_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_blups_pearson <- cnj04_blups_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj04_blups_ay_spearman <- cnj04_blups_ay_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_blups_ay_pearson <- cnj04_blups_ay_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    cnj04_blups_ny_spearman <- cnj04_blups_ny_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="spearman")
    cnj04_blups_ny_pearson <- cnj04_blups_ny_join %>% dplyr::select(blup.x,blup.y) %>% cor(use="pairwise.complete.obs",method="pearson")
    
```

## QTL Analyses

```{r qtl-table-defs, echo=FALSE, message=FALSE, include=TRUE}
#Copy qtl and effects tables to Data/publication/tables/ folder
file.copy(from=paste0(cnj02_workflow,"/traits/qtl_collated.consensus.csv"),to=paste0("../Data/publication/tables/qtl.cnj02.csv"),overwrite=TRUE)
file.copy(from=paste0(cnj02_workflow,"/traits/effects_collated.csv"),to=paste0("../Data/publication/tables/effects.cnj02.csv"),overwrite=TRUE)
file.copy(from=paste0(cnj04_workflow,"/traits/qtl_collated.consensus.csv"),to=paste0("../Data/publication/tables/qtl.cnj04.csv"),overwrite=TRUE)
file.copy(from=paste0(cnj04_workflow,"/traits/effects_collated.csv"),to=paste0("../Data/publication/tables/effects.cnj04.csv"),overwrite=TRUE)
qtls.cnj02.tb <- read_csv(paste0(cnj02_workflow,"/traits/effects_collated.csv"))
qtls.cnj02.s1.modv <- round(mean((qtls.cnj02.tb %>% filter(method=="scanone") %>% group_by(model) %>% summarize(model_variance=mean(model_variance)) %>% ungroup())$model_variance), digits=1)
qtls.cnj02.s1.markv <- round(mean((qtls.cnj02.tb %>% filter(method=="scanone") %>% group_by(model) %>% summarize(marker_variance=mean(marker_variance)) %>% ungroup())$marker_variance), digits=1)
qtls.cnj02.s1.nqtls <- round(mean((qtls.cnj02.tb %>% filter(method == "scanone") %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj02.s1.ntraits <- length(unique((qtls.cnj02.tb %>% filter(method == 'scanone') %>% group_by(model,trait) %>% summarize(marker_variance=mean(marker_variance)))$trait)) 
qtls.cnj02.sw.modv <- round(mean((qtls.cnj02.tb %>% filter(method=="stepwiseqtl") %>% group_by(model) %>% summarize(model_variance=mean(model_variance)) %>% ungroup())$model_variance), digits=1)
qtls.cnj02.sw.markv <- round(mean((qtls.cnj02.tb %>% filter(method=="stepwiseqtl") %>% group_by(model) %>% summarize(marker_variance=mean(marker_variance)) %>% ungroup())$marker_variance), digits=1)
qtls.cnj02.sw.nqtls <- round(mean((qtls.cnj02.tb %>% filter(method == "stepwiseqtl") %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj02.sw.ntraits <- length(unique((qtls.cnj02.tb %>% filter(method == 'stepwiseqtl'))$trait)) 
    
qtls.cnj02.collated.tb <- read_csv(paste0(cnj02_workflow,"/traits/qtl_collated.csv"),show_col_types = F) %>%
                            filter(GLRpvalue < 0.05) #Only include QTL from models with significant genotype effect (GLRpvalue < 0.05)
qtls.cnj02.collated.nints.tb <- qtls.cnj02.collated.tb %>%
                                filter(is.na(chr2) | 
                                       is.na(position2))
qtls.cnj02.collated.ints.tb <- qtls.cnj02.collated.tb %>%
                                filter(!is.na(chr2) &
                                      !is.na(position2))

qtls.cnj02.s1.collated.tb <- qtls.cnj02.collated.nints.tb %>%
                                filter(method=="scanone")
qtls.cnj02.sw.collated.tb <- qtls.cnj02.collated.nints.tb %>%
                                filter(method=="stepwiseqtl")

qtls.cnj02.s1.major.tb <- qtls.cnj02.s1.collated.tb %>%
                            filter(marker.variance >= 10)

qtls.cnj02.sw.major.tb <- qtls.cnj02.sw.collated.tb %>%
                            filter(marker.variance >= 10)

qtls.cnj02.s1.n <- nrow(qtls.cnj02.s1.collated.tb)
qtls.cnj02.s1.nmajor <- nrow(qtls.cnj02.s1.major.tb)
qtls.cnj02.sw.n <- nrow(qtls.cnj02.sw.collated.tb)
qtls.cnj02.sw.nmajor <- nrow(qtls.cnj02.sw.major.tb)

qtls.cnj04.collated.tb <- read_csv(paste0(cnj04_workflow,"/traits/qtl_collated.csv"),show_col_types = F) %>%
                            filter(GLRpvalue < 0.05) #Only include QTL from models with significant genotype effect (GLRpvalue < 0.05)
qtls.cnj04.collated.nints.tb <- qtls.cnj04.collated.tb %>%
                                filter(is.na(chr2) | 
                                       is.na(position2))
qtls.cnj04.collated.ints.tb <- qtls.cnj04.collated.tb %>%
                                filter(!is.na(chr2) &
                                      !is.na(position2))

qtls.cnj04.s1.collated.tb <- qtls.cnj04.collated.nints.tb %>%
                                filter(method=="scanone")
qtls.cnj04.sw.collated.tb <- qtls.cnj04.collated.nints.tb %>%
                                filter(method=="stepwiseqtl")

qtls.cnj04.s1.major.tb <- qtls.cnj04.s1.collated.tb %>%
                            filter(marker.variance >= 10)

qtls.cnj04.sw.major.tb <- qtls.cnj04.sw.collated.tb %>%
                            filter(marker.variance >= 10)

qtls.cnj04.s1.n <- nrow(qtls.cnj04.s1.collated.tb)
qtls.cnj04.s1.nmajor <- nrow(qtls.cnj04.s1.major.tb)
qtls.cnj04.sw.n <- nrow(qtls.cnj04.sw.collated.tb)
qtls.cnj04.sw.nmajor <- nrow(qtls.cnj04.sw.major.tb)

qtls.cnj02.sw.ints <- qtls.cnj02.collated.tb %>%
                        filter(method=="stepwiseqtl") %>%
                        filter(!is.na(chr2) & !is.na(position2))

qtls.cnj02.sw.ints.modv <- round(mean((qtls.cnj02.sw.ints %>% group_by(model) %>% summarize(model.variance=mean(model.variance)) %>% ungroup())$model.variance), digits=1)
qtls.cnj02.sw.ints.markv <- round(mean((qtls.cnj02.sw.ints %>% group_by(model) %>% summarize(marker.variance=mean(marker.variance)) %>% ungroup())$marker.variance), digits=1)
qtls.cnj02.sw.ints.nqtls <- round(mean((qtls.cnj02.sw.ints %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj02.sw.ints.ntraits <- length(unique((qtls.cnj02.sw.ints)$trait)) 
qtls.cnj02.sw.ay.ints <- qtls.cnj02.sw.ints %>%
                            filter(model=="all-years")
qtls.cnj02.sw.ay.nints <- nrow(qtls.cnj02.sw.ay.ints)
qtls.cnj02.sw.ay.ntraits <- length(unique(qtls.cnj02.sw.ay.ints$trait))
qtls.cnj02.sw.ay.markv <- round(mean(qtls.cnj02.sw.ay.ints$marker.variance, na.rm=T), digits=1)

qtls.cnj04.tb <- read_csv(paste0(cnj04_workflow,"/traits/effects_collated.csv"))
qtls.cnj04.s1.modv <- round(mean((qtls.cnj04.tb %>% filter(method=="scanone") %>% group_by(model) %>% summarize(model_variance=mean(model_variance)) %>% ungroup())$model_variance), digits=1)
qtls.cnj04.s1.markv <- round(mean((qtls.cnj04.tb %>% filter(method=="scanone") %>% group_by(model) %>% summarize(marker_variance=mean(marker_variance)) %>% ungroup())$marker_variance), digits=1)
qtls.cnj04.s1.nqtls <- round(mean((qtls.cnj04.tb %>% filter(method == "scanone") %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj04.s1.ntraits <- length(unique((qtls.cnj04.tb %>% filter(method == 'scanone') %>% group_by(model,trait) %>% summarize(marker_variance=mean(marker_variance)))$trait)) 
qtls.cnj04.sw.modv <- round(mean((qtls.cnj04.tb %>% filter(method=="stepwiseqtl") %>% group_by(model) %>% summarize(model_variance=mean(model_variance)) %>% ungroup())$model_variance), digits=1)
qtls.cnj04.sw.markv <- round(mean((qtls.cnj04.tb %>% filter(method=="stepwiseqtl") %>% group_by(model) %>% summarize(marker_variance=mean(marker_variance)) %>% ungroup())$marker_variance), digits=1)
qtls.cnj04.sw.nqtls <- round(mean((qtls.cnj04.tb %>% filter(method == "stepwiseqtl") %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj04.sw.ntraits <- length(unique((qtls.cnj04.tb %>% filter(method == 'stepwiseqtl'))$trait)) 
qtls.cnj04.sw.ints <- read_csv(paste0(cnj04_workflow,"/traits/qtl_collated.csv")) %>%
                            filter(method=="stepwiseqtl") %>%
                            filter(!is.na(chr2) & !is.na(position2))
qtls.cnj04.sw.ints.modv <- round(mean((qtls.cnj04.sw.ints %>% group_by(model) %>% summarize(model.variance=mean(model.variance)) %>% ungroup())$model.variance), digits=1)
qtls.cnj04.sw.ints.markv <- round(mean((qtls.cnj04.sw.ints %>% group_by(model) %>% summarize(marker.variance=mean(marker.variance)) %>% ungroup())$marker.variance), digits=1)
qtls.cnj04.sw.ints.nqtls <- round(mean((qtls.cnj04.sw.ints %>% nest_by(model,trait,chr,position) %>% group_by(model,trait) %>% summarize(nqtls=n()))$nqtls), digits=1) 
qtls.cnj04.sw.ints.ntraits <- length(unique((qtls.cnj04.sw.ints)$trait)) 
qtls.cnj04.sw.ay.ints <- qtls.cnj04.sw.ints %>%
                            filter(model=="all-years")
qtls.cnj04.sw.ay.nints <- nrow(qtls.cnj04.sw.ay.ints)
qtls.cnj04.sw.ay.ntraits <- length(unique(qtls.cnj04.sw.ay.ints$trait))
qtls.cnj04.sw.ay.markv <- round(mean(qtls.cnj04.sw.ay.ints$marker.variance, na.rm=T), digits=1)
```

### CNJ02

For the CNJ02 population, we discovered an average of
`r qtls.cnj02.s1.nqtls` QTL per *trait × model* combination across
`r qtls.cnj02.s1.ntraits` traits for the `scanone` method and
`r qtls.cnj02.sw.nqtls` QTL per *trait × model* across
`r qtls.cnj02.sw.ntraits` traits for the `stepwiseqtl` method. The
average `scanone` variance per QTL was `r qtls.cnj02.s1.markv` percent,
with an average model variance of `r qtls.cnj02.s1.modv` percent. The
`mean stepwiseqtl` variance per QTL was `r qtls.cnj02.sw.markv` percent,
with an average model variance of `r qtls.cnj02.sw.modv` percent. A more
detailed examination of trait QTL, effect plots, LOD profiles, and trait
correlations can be interrogated on the [CNJ02 QTL
portal](https://vaccinium.vcru.wisc.edu/circos-cnj02/) webapp.

<!-- A summary of the top `stepwiseqtl` QTL discovered for the CNJ02 -->
<!-- population for the *all years* model can be found in Table -->
<!-- \@ref(tab:qtl-table-cnj02-sw). This summary of QTL only shows trait -->
<!-- models with significant genotype effects and heritability higher than -->
<!-- 0.5, sorted from highest to lowest heritability (same trait order as -->
<!-- \@ref(tab:blups-table-cnj02-all-top)). From the marker variance and -->
<!-- effect plots, a few traits and their associated QTL of note emerge. For -->
<!-- largest upright *berry length*, linkage group 10 at position 51.79cM, -->
<!-- progeny display a 1.94mm advantage of allele *B* over *A* (maternal -->
<!-- effect) and a 1.09mm advantage of *C* over *D* (paternal effect), with -->
<!-- progeny genotype *BC* showing the largest breeding advantage. A similar -->
<!-- pattern emerges with the largest berry *number of seeds* on linkage -->
<!-- group 6 (position 51.28cM), although the *BC* genotype advantage is not -->
<!-- as salient due to larger genotype variance and a lower mean interaction -->
<!-- effect. For *total anthocyanins* (Tacy), linkage group 3, position -->
<!-- 56.18cM shows a clear advantage in the *·D* genotype over *·C* genotype, -->
<!-- with an average higher Tacy content of around 12.6 mg units per 100 mg -->
<!-- fruit. As opposite sides of the same coin, *sound fruit yield* and -->
<!-- *percent fruit rot* have a collocated QTL on LG 11 (21.88cM and 17.25cM, -->
<!-- respectively) that appears to be have a noticeable impact on fruit rot -->
<!-- in CNJ02. CNJ02 progeny with genotype *AC* shows evidence of higher -->
<!-- *sound fruit yield* and lower *percent fruit rot* than the other -->
<!-- genotypes. Largest *berry weight* and *upright mean fruit mass* both -->
<!-- have a QTL on LG 11 at position 30.32cM, with the average effect of -->
<!-- allele *D* giving an average 0.42 gram and 0.36 gram/berry advantage -->
<!-- over allele *C*. Plot *total yield* demonstrates and average allelic -->
<!-- effect of *C* over *D* of around 65 grams per square foot on linkage -->
<!-- group 12 (position 48.76cM). Given that the genotypic effects for the -->
<!-- upright berry *length versus width* were not found to be significant in -->
<!-- the *all-year* model for CNJ02, the berry *chimera length versus width* -->
<!-- and *chimera eccentricity* represent good proxies. A collocated QTL on -->
<!-- linkage group 3 near 30cM for traits *chimera length versus width* and -->
<!-- *chimera eccentricity* shows a small, but statisically meaningful -->
<!-- association such that genotype *BD* will produce rounder berries, an -->
<!-- important characteristic for fruit processing. -->

```{r qtl-table-cnj02-sw-setup, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE}
load(file=paste0(cnj02_workflow,"/.RData.01_genBLUP"))
models.filtered <- model.collated.summary.wide.tb %>% 
                        filter(model == 'all-years') %>% 
                        filter(h2 > cnj0x.median.h2.ay.r) %>% 
                        filter(GLRpvalue < 0.05) %>% 
                        arrange(desc(h2)) %>%
                        mutate(trait=factor(as.character(trait), levels=as.character(trait), ordered=TRUE))
```

```{r qtl-table-cnj02-sw, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption <- paste0("CNJ02 Population QTL for `all years` model, QTL fit using stepwiseqtl() function, with effect plots.  Only models with heritability > ",cnj0x.median.h2.ay.r," and with significant genotypic effects ($p < 0.05$) are displayed.")
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
if( is_pdf_output() ) {
    caption = knitr:::escape_latex(caption)
}
load(file=paste0(cnj02_workflow,"/.RData.12_geneffects.2"))
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

tbl <- generateReducedTableNoPlots((effs.complete.tb %>% 
                     filter(model=='all-years' &
                            method=='stepwiseqtl' &
                            !is.na(nearest_marker) &
                            trait %in% models.filtered$trait) %>% 
                     group_by(trait,model) %>%
                     mutate(top_qtls = c(rep(TRUE,min(1,length(marker_variance))),rep(FALSE,length(marker_variance)-min(1,length(marker_variance))))) %>%
                     filter(top_qtls == TRUE) %>%
                     dplyr::select(-top_qtls) %>%
                     ungroup() %>%
                     mutate(trait = factor(as.character(trait), levels=levels(models.filtered$trait), ordered=TRUE)) %>%
                     arrange(trait,desc(marker_variance)))[1:6,],
                     caption=caption,
                     tfont_size=table.font_size)
print_table(tbl, clabel, caption)
save_as_docx(tbl,path="../Data/publication/tables/qtl-table-cnj02-sw.docx")
```
<!-- The `stepwiseqtl` method found an average of -->
<!-- `r qtls.cnj02.sw.ints.nqtls` CNJ02 QTL:QTL interactions per *trait × -->
<!-- model* combination across `r qtls.cnj02.sw.ints.ntraits` traits, with an -->
<!-- average model variance of `r qtls.cnj02.sw.ints.modv` percent and an -->
<!-- average variance per QTL pair of `r qtls.cnj02.sw.ints.markv` percent. -->
<!-- The *all years* model found `r qtls.cnj02.sw.ay.nints` significant -->
<!-- pairwise QTL:QTL interactions across `r qtls.cnj02.sw.ay.ntraits` -->
<!-- traits, with an average variance per QTL pair of -->
<!-- `r qtls.cnj02.sw.ay.markv` percent. Table -->
<!-- \@ref(tab:qtl-table-cnj02-sw-ints) summarizes the top significant -->
<!-- pairwise QTL interactions for the *all years* model. The boxplots -->
<!-- indicate the separate BLUP distributions grouped by pairwise QTL:QTL -->
<!-- genotypes (subplot a), the first QTL genotypes (subplot c), and the -->
<!-- second QTL genotypes (subplot b). For the pairwise QTL:QTL genotype -->
<!-- boxplots (subplot a), black diamond markers indicate the mean of the sum -->
<!-- of the two QTL if considering their additive effect sizes only, while -->
<!-- the black circle markers indicate the mean pairwise QTL:QTL effect. The -->
<!-- horizontal distance, or bias, between these means visually highlights -->
<!-- the additional difference a QTL:QTL interaction imparts on the additive -->
<!-- effects of the two separate QTL. The *MAD* metric, or *maximum average -->
<!-- distance*, takes the maximum of all these biases, and *NMAD* represents -->
<!-- the normalized value of the *MAD* metric. *MAD* highlights the most -->
<!-- extreme value an QTL:QTL interaction can impart on the additive effect, -->
<!-- while *NMAD* gives a relative metric for comparing across traits with -->
<!-- different units. -->

<!-- *Total anthocyanins* has the largest *NMAD* metric, with a maximum mean -->
<!-- epistatic effect size of 4.591 $\frac{mg}{100g\ berries}$ due to a -->
<!-- pairwise QTL interaction found on linkage groups 4 and 7. *Number of -->
<!-- seeds* is a close second, with a maximum mean epistatic effect size of -->
<!-- 2.412 from a QTL interaction on linkages groups 5 and 6. Other -->
<!-- interactions of note include those found for traits upright largest -->
<!-- *berry weight* (*MAD*≈0.141g), *upright mean fruit mass* -->
<!-- (*MAD*≈0.114g/berry), and upright largest *berry width* -->
<!-- (*MAD*≈0.164mm)(\@ref(tab:qtl-table-cnj02-sw-ints)). -->

<!-- ```{r qtl-table-cnj02-sw-ints, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", eval=FALSE} -->
<!-- clabel <- knitr::opts_current$get('label') -->
<!-- caption="CNJ02 Population Top Interaction QTL for all years model, QTL fit using stepwiseqtl() function.  Only models with heritability > 0.5, QTL with interaction percent variance explained greater than 5.0 percent, and models with significant genotypic effects ($p < 0.05$) are displayed." -->
<!-- load(file=paste0(cnj02_workflow,"/.RData.13_geninteffects.2")) -->
<!-- if(is.null(getOption("knitr.in.progress"))) { -->
<!--     source("R/replaceKnitrPandocFormats.R") -->
<!-- } -->
<!-- if( is_pdf_output() ) { -->
<!--     caption = knitr:::escape_latex(caption) -->
<!-- } -->
<!-- tbl <- generateReducedTableNoPlots(effs.ints.all.tb %>%  -->
<!--                      filter(model=='all-years' & -->
<!--                             trait %in% models.filtered$trait, -->
<!--                             marker_variance >= 5.0) %>%  -->
<!--                     mutate(trait = factor(as.character(trait), levels=levels(models.filtered$trait), ordered=TRUE)) %>% -->
<!--                     group_by(trait) %>% -->
<!--                     arrange(desc(marker_variance), .by_group = TRUE) %>% -->
<!--                     slice_head() %>% -->
<!--                     ungroup(), -->
<!--                     caption=caption, -->
<!--                     tfont_size=table.font_size) -->
<!-- print_table(tbl, clabel, caption) -->
<!-- save_as_docx(tbl,path="../Data/publication/tables/qtl-table-cnj02-sw-ints.docx") -->
<!-- ``` -->

<!-- For yield-relevant traits, three linkage groups stand out: linkage -->
<!-- groups 3, 10, and 11 (figure \@ref(fig:qtl-plot-cnj02)A). The greatest -->
<!-- set of collocated QTL is found on linkage group 3, with largest *berry -->
<!-- length*, largest *berry weight*, *plot mean fruit mass*, and *upright -->
<!-- mean fruit mass* all found near 46-47cM. Linkage group 10 is noteable -->
<!-- for having the highest number of large-effect yield QTL. On linkage -->
<!-- group 10, largest *berry length* and largest *berry weight* are closely -->
<!-- linked (52cM and 49cM, respectively), as are *plot mean fruit mass* and -->
<!-- *upright mean fruit mass* (39cM and 37cM, respectively). Also within -->
<!-- linkage group 10, large and signficant QTL can be found for largest -->
<!-- *berry width* (28cM), *length vs. width* (63cM), and upright *total -->
<!-- berry weight* (43cM). On linkage group 11, largest *berry weight* and -->
<!-- *upright mean fruit mass* are collocated at 30cM, while plot mean fruit -->
<!-- mass is closely linked at 25cM. QTL for largest *berry width* (40cM) and -->
<!-- upright *length vs. width* (81cM) can also be found on linkage group 11. -->

<!-- Berry chemistry trait QTL are highlighted in figure -->
<!-- \@ref(fig:qtl-plot-cnj02)B. There are no clear collocation patterns that -->
<!-- emerge as with the yield-related traits, not surprising given that the -->
<!-- chemistry trait correlations were not as strong as those of yield -->
<!-- traits. Some of the traits and their QTL were discussed in reference to -->
<!-- table \@ref(tab:qtl-table-cnj02-sw). A few non-chemistry traits cohabit -->
<!-- chemistry traits in figure \@ref(fig:qtl-plot-cnj02)B because of their -->
<!-- correlation with corresponding chemistry traits. For instance, *percent -->
<!-- fruit rot* and *Tacy* have a moderately positive correlation (0.38), -->
<!-- likely related to large QTL closely linked on linkage group 7 (position -->
<!-- 38cM and 33cM, respectively). These two traits share the same *BD* -->
<!-- genotype for the highest positive effect. The major QTL for *Tacy* found -->
<!-- on linkage group 3 is also linked to a large QTL for *plot mean fruit -->
<!-- mass*, with allele *D* versus *C* showing a strong positive *Tacy* -->
<!-- effect and substantial negative *plot mean fruit mass* effect (table -->
<!-- \@ref(tab:qtl-table-cnj02-sw)). The largest QTL for *titratable -->
<!-- acidity*, found on linkage group 2 at 50cM, appear to be closely linked -->
<!-- with other traits' major QTL. A closely linked pair of QTL on linkage -->
<!-- group 3 between *plot mean fruit mass* and *titratable acidity* could -->
<!-- explain the modest positive correlation between these two traits (0.40) -->
<!-- (TODO - include correlation table supplemental reference). -->

<!-- ```{r qtl-plot-cnj02-setup, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE} -->
<!--     imageA <- "../Data/publication/figures/cnj02.circos.yield.png" -->
<!--     imageAA <- "../Data/publication/figures/cnj02.circos.yield.A.png" -->
<!--     imageB <- "../Data/publication/figures/cnj02.circos.chemistry.png" -->
<!--     imageBB <- "../Data/publication/figures/cnj02.circos.chemistry.B.png" -->
<!--     imageCombined <- "../Data/publication/figures/cnj02.circos.png" -->
<!-- ``` -->

<!-- ```{r qtl-plot-cnj02-annotate, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE} -->
<!--     system2("misc/annotatefig.sh", args=c(imageA,"A")) -->
<!--     system2("misc/annotatefig.sh", args=c(imageB,"B")) -->
<!-- ``` -->

<!-- ```{r qtl-plot-cnj02-montage, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE} -->
<!--     system2("montage", args=c("-background","white","-geometry","2048x2048+20+0","-tile","2x1",imageAA,imageBB,imageCombined)) -->
<!-- ``` -->

<!-- ```{r qtl-plot-cnj02, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Notable QTL found in population CNJ02.  A) Yield-related traits.  B) Chemistry and chemistry-correlated traits.", eval=FALSE} -->
<!--     knitr::include_graphics(imageCombined) -->
<!-- ``` -->

<!-- ### CNJ04 -->

<!-- Despite having a lower population size (*n = 72*), and thus lower power -->
<!-- to find statistically significant QTL, QTL were still found in CNJ04. -->
<!-- The CNJ04 population showed average of `r qtls.cnj04.s1.nqtls` QTL per -->
<!-- *trait × model* combination across `r qtls.cnj04.s1.ntraits` traits for -->
<!-- the `scanone` method and `r qtls.cnj04.sw.nqtls` QTL per *trait × model* -->
<!-- across `r qtls.cnj04.sw.ntraits` traits for the `stepwiseqtl` method. -->
<!-- The `scanone` average total BLUP variance explained per QTL was -->
<!-- `r qtls.cnj04.s1.markv` percent, with an average BLUP variance explained -->
<!-- by all QTL per *trait × model* of `r qtls.cnj04.s1.modv` percent. The -->
<!-- `stepwiseqtl` average total BLUP variance explained per QTL was -->
<!-- `r qtls.cnj04.sw.markv` percent, with an average BLUP variance explained -->
<!-- by all QTL per *trait × model* of `r qtls.cnj04.sw.modv` percent. A more -->
<!-- detailed investigation of trait QTL, effect plots, LOD profiles, and -->
<!-- trait correlations can be explored through the [CNJ04 QTL -->
<!-- portal](https://vaccinium.vcru.wisc.edu/circos-cnj04/) webapp. -->

<!-- Similar to population CNJ02, a summary of the top 2 `stepwiseqtl` QTL -->
<!-- for the *all years* model can be found in Table -->
<!-- \@ref(tab:qtl-table-cnj04-sw), with traits sorted by decreasing -->
<!-- heritability (same trait order as \@ref(tab:blups-table-cnj04-all-top)). -->
<!-- For largest upright berry *length vs. width*, linkage group 11, position -->
<!-- 11.74cM, progeny with *BD* genotype demonstrate around 10 percent more -->
<!-- roundness in berry shape. A highly linked QTL in largest *berry length*, -->
<!-- linkage group 11 and position 5.09cM, appears to be the main driver of -->
<!-- this lower *length vs. width* ratio. For *Tacy*, although a QTL on -->
<!-- linkage group 6 explains most of the genetic variation (10.08 percent) -->
<!-- in the trait, linkage group 3 shows evidence of having an important role -->
<!-- in explaining *Tacy* variation just as in the CNJ02 population. This QTL -->
<!-- on linkage group 3, position 36.49cM, suggests that most of the -->
<!-- variation is on the paternal side, with a 14.21 -->
<!-- $\frac{mg}{100mg\ fruit}$ advantage of allele *D* over *C* (figure -->
<!-- \@ref(fig:qtl-plot-cnj04)B).\ -->
<!-- Also as in CNJ02, the largest significant QTL for *total yield* is found -->
<!-- on linkage group 12 and position 48.76cM, with genotype *AC* showing the -->
<!-- largest genotype effect. Interestingly, despite having a moderate -->
<!-- positive correlation with largest berry *length vs. width* (0.62), the -->
<!-- top two QTL for *chimera length v. width* are on linkage groups 4 -->
<!-- (10.82cM) and 1 (14.48cM). These QTL are closely linked to other -->
<!-- significant QTL found for *length vs. width* without appearing to be -->
<!-- linked to QTL associated with *berry width* or *berry length*, avoiding -->
<!-- negative linkage drag ([supplemental link](#cnj04-qtl)). If a lower -->
<!-- overall acidity is preferred, a significant QTL at linkage group 11, -->
<!-- position 16.42cM, suggests thatdplyr::selecting genotype *BD* will help lower -->
<!-- acidity in subsequent generations. If wanting todplyr::select for higher -->
<!-- *proanthocyanins* due to their beneficial effects on human health, a QTL -->
<!-- on linkage group 7, position 25.82cM, indicates a modest improvement by -->
<!-- selecting *AC*. Although the absolute effect size is modest and the -->
<!-- model variance explained by the QTL is small, *Brix* shows a modest -->
<!-- advantage in genotype *BC* (around 0.2 percent) on linkage group 9, -->
<!-- position 84.81cM. The very high correlation between *total yield* and -->
<!-- *sound fruit yield* (0.96) is consistent with these two traits sharing -->
<!-- the same top 2 QTL. -->

<!-- For yield-relevant traits, linkage group 10 in the CNJ04 population has -->
<!-- the largest number of top-tier QTL (figure \@ref(fig:qtl-plot-cnj04)A). -->
<!-- QTL of note on linkage group 10 not already mentioned include collocated -->
<!-- *total yield* and *plot mean fruit mass* (position 61cM) and the closely -->
<!-- linked upright trait *total berry weight* (position 65cM). These closely -->
<!-- linked QTL likely explain the modestly positive correlations (\~0.5) -->
<!-- between these traits. Linkage group 9 has collocated QTL for traits -->
<!-- largest *berry weight* and upright *upright mean fruit mass* (position -->
<!-- 58cM), closely linked with a large QTL for largest *berry weight* -->
<!-- (position 64cM). -->

```{r qtl-table-cnj04-sw-setup, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE}
load(file=paste0(cnj04_workflow,"/.RData.01_genBLUP"))
models.filtered <- model.collated.summary.wide.tb %>% 
                        filter(model == 'all-years') %>% 
                        filter(h2 > 0.5) %>% 
                        filter(GLRpvalue < 0.05) %>% 
                        arrange(desc(h2)) %>%
                        mutate(trait=factor(as.character(trait), levels=as.character(trait), ordered=TRUE))
```

```{r qtl-table-cnj04-sw, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption <- "CNJ04 Population Top 2 QTL for all years model, QTL fit using stepwiseqtl() function, with effect plots.  Only models with $heritability > 0.5$ and with significant genotypic effects ($p < 0.05$) are displayed."
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
if( is_pdf_output() ) {
    caption = knitr:::escape_latex(caption)
}
load(file=paste0(cnj04_workflow,"/.RData.12_geneffects.2"))
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
tbl <- generateReducedTableNoPlots((effs.complete.tb %>% 
                     filter(model=='all-years' &
                            method=='stepwiseqtl' &
                            !is.na(nearest_marker) &
                            trait %in% models.filtered$trait) %>% 
                     mutate(trait = factor(as.character(trait), levels=levels(models.filtered$trait), ordered=TRUE)) %>%
                     group_by(trait,model) %>%
                    mutate(top_qtls = c(rep(TRUE,min(1,length(marker_variance))),rep(FALSE,length(marker_variance)-min(1,length(marker_variance))))) %>%
                    filter(top_qtls == TRUE) %>%
                    dplyr::select(-top_qtls) %>%
                    ungroup() %>%
                     arrange(trait,desc(marker_variance)))[1:6,],
                     caption = caption,
                     tfont_size=table.font_size)
print_table(tbl, clabel, caption)
save_as_docx(tbl,path="../Data/publication/tables/qtl-table-cnj04-sw.docx")
```


<!-- An average of `r qtls.cnj04.sw.ints.nqtls` CNJ04 QTL:QTL interactions -->
<!-- per *trait × model* combination across `r qtls.cnj04.sw.ints.ntraits` -->
<!-- traits, with an average model variance of `r qtls.cnj04.sw.ints.modv` -->
<!-- percent and an average variance per QTL pair of -->
<!-- `r qtls.cnj04.sw.ints.markv` percent. The *all years* model found -->
<!-- `r qtls.cnj04.sw.ay.nints` significant pairwise QTL:QTL interactions -->
<!-- across `r qtls.cnj04.sw.ay.ntraits` traits, with an average variance per -->
<!-- QTL pair of `r qtls.cnj04.sw.ay.markv` percent. Table -->
<!-- \@ref(tab:qtl-table-cnj04-sw-ints) summarizes the top significant -->
<!-- pairwise QTL interactions for the *all years* model, with the same -->
<!-- semantics of the subplots as defined for -->
<!-- \@ref(tab:qtl-table-cnj02-sw-ints). The number of pairwise QTL -->
<!-- interactions for the CNJ04 population was considerably lower than that -->
<!-- of CNJ02 (`r qtls.cnj04.sw.ay.nints` vs. `r qtls.cnj02.sw.ay.nints`; -->
<!-- predictable given its much smaller population size. None of the highest -->
<!-- heritability traits with significant model genotype effects had -->
<!-- interaction variances above 5 percent. Only one interaction is featured -->
<!-- in table \@ref(tab:qtl-table-cnj04-sw-ints) for QTL mapped across all -->
<!-- years: *Brix*. With a *MAD* metric of only 0.129°, this pairwise -->
<!-- interaction on linkages groups 6 and 11 would not likely pass on any -->
<!-- meaningful effect on berry taste and quality. -->

```{r qtl-table-cnj04-sw-ints, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption="CNJ04 Population Top Interaction QTL for all years model, QTL fit using stepwiseqtl() function.  Only models with $heritability > 0.5$ and with significant genotypic effects ($p < 0.05$) are displayed."
load(file=paste0(cnj04_workflow,"/.RData.13_geninteffects.2"))
if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}
if( is_pdf_output() ) {
    caption = knitr:::escape_latex(caption)
}
tbl <- generateReducedTableNoPlots(effs.ints.all.tb %>% 
                    filter(model=='all-years' &
                           trait %in% models.filtered$trait) %>% 
                    mutate(trait = factor(as.character(trait), levels=levels(models.filtered$trait), ordered=TRUE)) %>%
                    group_by(trait) %>%
                    arrange(desc(marker_variance), .by_group = TRUE) %>%
                    slice_head() %>%
                    ungroup(),
                    caption=caption,
                    tfont_size=table.font_size)
print_table(tbl, clabel, caption)
save_as_docx(tbl,path="../Data/publication/tables/qtl-table-cnj04-sw-ints.docx")
```

```{r qtl-plot-cnj04-setup, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    imageA <- "../Data/publication/figures/cnj04.circos.yield.png"
    imageAA <- "../Data/publication/figures/cnj04.circos.yield.A.png"
    imageB <- "../Data/publication/figures/cnj04.circos.chemistry.png"
    imageBB <- "../Data/publication/figures/cnj04.circos.chemistry.B.png"
    imageCombined <- "../Data/publication/figures/cnj04.circos.png"
```

```{r qtl-plot-cnj04-annotate, echo=FALSE, message=FALSE, include=FALSE, eval=FALSE}
    system2("misc/annotatefig.sh", args=c(imageA,"A"))
    system2("misc/annotatefig.sh", args=c(imageB,"B"))
```

```{r qtl-plot-cnj04-montage, echo=FALSE, message=FALSE, include=TRUE, eval=FALSE}
    system2("montage", args=c("-background","white","-geometry","2048x2048+20+0","-tile","2x1",imageAA,imageBB,imageCombined))
```

```{r qtl-plot-cnj04, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Notable QTL found in population CNJ04.  A) Yield-related traits.  B) Chemistry and chemistry-correlated traits.", eval=FALSE}
    knitr::include_graphics(imageCombined)
```

### Co-located QTL

```{r qtl-table-cnj0x-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis'}
generateColocatedQTLTable <- function(tbl) {
    if(is.null(getOption("knitr.in.progress"))) {
        source("R/replaceKnitrPandocFormats.R")
    }
    if ( is_word_output() ) {
        ftbl <- tbl %>%
                    mutate(model=unlist(lapply(str_split(model, fixed('+')), function(x) { paste0(sort(x),collapse='+\n') })),
                           method=unlist(lapply(str_split(method, fixed('+')), function(x) { paste0(sort(x),collapse='+\n') }))) %>%
                    flextable(col_keys=c("trait","model","chr","position","mean_marker_variance","interval_left","interval_right","method","mean_AvB","mean_CvD","mean_Int")) %>%
                    set_header_labels(trait="Trait",
                                      model="Models",
                                      chr="LG",
                                      position="Position (cM)",
                                      mean_marker_variance="R^2^",
                                      interval_left="-1.5 LOD (cM)",
                                      interval_right="+1.5 LOD (cM)",
                                      method="Methods") %>%
                    ftExtra::colformat_md(mean_marker_variance, part="header") %>%
                    fontsize(part="body",
                            size=11) %>%
                    fontsize(part="header",
                             size = 12) %>%
                    mk_par(j =~mean_AvB, value=as_paragraph(as_equation("\\mathbf{\\overline{AvB}}")), part="header") %>%
                    mk_par(j =~mean_CvD, value=as_paragraph(as_equation("\\mathbf{\\overline{CvD}}")), part="header") %>%
                    mk_par(j =~mean_Int, value=as_paragraph(as_equation("\\mathbf{\\overline{Int}}")), part="header") %>%
                    flextable::footnote(i = 1, 
                            j = ~position+mean_marker_variance+method+interval_left+interval_right+mean_AvB+mean_CvD+mean_Int,
                            value = as_paragraph(c("Mean position of combined QTL",
                                                 "Mean of variance explained by combined QTL",
                                                 "QTL mapping method applied",
                                                 "1.5 LOD left interval",
                                                 "1.5 LOD right interval",
                                                 "Mean maternal effect size: (AC+AD)-(BC+BD)",
                                                 "Mean paternal effect size: (AC+BC)–(AD+BD)",
                                                 "Mean interaction effect size: (AC+BD)-(AD+BC)")),
                            ref_symbols = c(" a"," b"," c", " d", " e", " f", " g", " h"),
                                    part = "header") %>%
                    bold(part="header") %>%
                    bold(j =~ trait, part="body") %>%
                    fontsize(part="footer",
                             size=10) %>%
                    set_formatter(mean_AvB = function(x) { signif.digits.char(x, digits = 2) }) %>%
                    set_formatter(mean_CvD = function(x) { signif.digits.char(x, digits = 2) }) %>%
                    set_formatter(mean_Int = function(x) { signif.digits.char(x, digits = 2) }) %>%
                    colformat_double(j=~position+mean_marker_variance+interval_left+interval_right, digits=1)
                    
                    
    } else {
    }
    return(ftbl)
}
```

#### CNJ02

```{r qtl-table-cnj02-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="Co-located, stable QTL in cranberry population CNJ02 found in at least three of four BLUP models.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance."}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj02 <- read_csv("../Data/publication/tables/cnj02_qtl_collated.consensus.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
fcolocated_cnj02 <- generateColocatedQTLTable(colocated_cnj02)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj02, path="../Data/publication/tables/cnj02_qtl_collated.consensus.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj02, clabel, caption)
```
```{r qtl-table-cnj02-nayay-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="Co-located QTL in cranberry population CNJ02 found in at least two of three years and in 'all-years' model.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance.", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj02 <- read_csv("../Data/publication/tables/cnj02_qtl_collated.consensus.nayay.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
fcolocated_cnj02 <- generateColocatedQTLTable(colocated_cnj02)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj02, path="../Data/publication/tables/cnj02_qtl_collated.consensus.nayay.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj02, clabel, caption)
```

```{r qtl-table-cnj02-nay-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="QTL in population CNJ02 found in at least two of three years.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance.", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj02_nay <- read_csv("../Data/publication/tables/cnj02_qtl_collated.consensus.nay.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
fcolocated_cnj02_nay <- generateColocatedQTLTable(colocated_cnj02_nay)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj02_nay, path="../Data/publication/tables/cnj02_qtl_collated.consensus.nay.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj02_nay, clabel, caption)
```

```{r qtl-fig-cnj02-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Co-located, stable QTL in cranberry population CNJ02 found in at least three of four BLUP models.  Only QTL with R^2^ ≥ 0.1 are shown. Shaded bars represent the 1.5LOD confidence interval for the original QTL, with the darker line segments indicating the location of the QTL.  The linkage groups are divided into four columns, one for each fitted BLUP model -- — from left to right: 2011, 2012, 2013, and all-years.  Each QTL interval is color-coded by its corresponding trait.  The labels and interval lines to the right of linkage groups show the stable meta-QTL associated with their respective genomic regions."}
devtools::load_all("~/software/bliptrip/LinkageMapView")
load('R/.RData.16_plotcolocatedqtllinkagemaps.R')
colocated_cnj02 <- read_csv("../Data/publication/tables/cnj02_qtl_collated.consensus.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
oldpar <- generate_linkagemapplots_primitive(colocated_cnj02, all_qtl, "../Data/publication/figures/cnj02_qtl.consensus.png", genetic_map_lg_summary, sectcolindex="model", lg.col="aliceblue", lg.minheight=5.3, lgperrow=5, noloci=T, pdf.width=6.8, alpha=0.5, jitterloci=NULL, render.type="png")
par(oldpar)
knitr::include_graphics("../Data/publication/figures/cnj02_qtl.consensus.png")
```

#### CNJ04

```{r qtl-table-cnj04-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="Co-located, stable QTL in cranberry population CNJ04 found in at least three of four BLUP models.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance."}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj04 <- read_csv("../Data/publication/tables/cnj04_qtl_collated.consensus.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
fcolocated_cnj04 <- generateColocatedQTLTable(colocated_cnj04)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj04, path="../Data/publication/tables/cnj04_qtl_collated.consensus.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj04, clabel, caption)
```

```{r qtl-table-cnj04-nayay-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="QTL in population CNJ04 found in at least two of three years and 'all-years' model.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance.", eval=FALSE}
#Nayay stands for 'not all years' and 'all-years', and means that we have at least 2/3 of year model QTL overlapping with an all-years model QTL
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj04_nayay <- read_csv("../Data/publication/tables/cnj04_qtl_collated.consensus.nayay.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
fcolocated_cnj04_nayay <- generateColocatedQTLTable(colocated_cnj04_nayay)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj04_nayay, path="../Data/publication/tables/cnj04_qtl_collated.consensus.nayay.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj04_nayay, clabel, caption)
```

```{r qtl-table-cnj04-nay-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="85%", tab.cap="QTL in population CNJ04 found in at least two of three years.  Only QTL with R^2^ ≥ 0.1 are shown.  Table is arranged in descending order by mean marker variance.", eval=FALSE}
clabel <- knitr::opts_current$get('label')
caption <- knitr::opts_current$get('tab.cap')

if(is.null(getOption("knitr.in.progress"))) {
    source("R/replaceKnitrPandocFormats.R")
}

colocated_cnj04_nay <- read_csv("../Data/publication/tables/cnj04_qtl_collated.consensus.nay.filtered.csv", show_col_types = FALSE)
fcolocated_cnj04_nay <- generateColocatedQTLTable(colocated_cnj04_nay)
if(is.null(getOption("knitr.in.progress"))) {
    if ( is_word_output() ) {
        flextable::save_as_docx(fcolocated_cnj04_nay, path="../Data/publication/tables/cnj04_qtl_collated.consensus.nay.filtered.docx", 
                                pr_section=prop_section(page_size = page_size(orient="landscape"), type="continuous"),
                                align="center")
    } else {
    }
}
print_table(fcolocated_cnj04_nay, clabel, caption)
```

```{r qtl-fig-cnj04-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Co-located QTL in cranberry population CNJ04 found in at least three of four BLUP models.  Only QTL with R^2^ ≥ 0.1 are shown. Shaded bars represent the 1.5LOD confidence interval for the original QTL, with the darker line segments indicating the location of the QTL.  The linkage groups are divided into four columns, one for each fitted BLUP model -- — from left to right: 2011, 2012, 2014, and all-years.  Each QTL interval is color-coded by its corresponding trait.  The labels and interval lines to the right of linkage groups show the stable meta-QTL associated with their respective genomic regions."}
devtools::load_all("~/software/bliptrip/LinkageMapView")
load('R/.RData.16_plotcolocatedqtllinkagemaps.R')
colocated_cnj04 <- read_csv("../Data/publication/tables/cnj04_qtl_collated.consensus.filtered.csv", show_col_types = FALSE) %>%
                            filter(trait %in% cnj0x.h2.top.nayay$trait)
oldpar <- generate_linkagemapplots_primitive(colocated_cnj04, all_qtl, "../Data/publication/figures/cnj04_qtl.consensus.png", genetic_map_lg_summary, lg.col="aliceblue", sectcolindex="model", lg.minheight=5.3, lgperrow=5, noloci=T, alpha=0.5)
par(oldpar)
knitr::include_graphics("../Data/publication/figures/cnj04_qtl.consensus.png")
```

#### CNJ0x

Figure \@ref(fig:qtl-fig-cnj0x-grouped-colocated) shows the principal QTL that compose the collocated QTL for sets CNJ02, CNJ04, and CNJ0x, as listed in Table 2.4. Together they represent stable QTL regions derived from primary QTL found in the current study for populations CNJ02, CNJ04, and CNJ0x (CNJ02 and CNJ04 combined). The traits that pass the filtering criteria outlined in Table 2.4 are all yield or yield-adjacent traits. Composite QTL are found on linkage groups 2, 3, 10, 11, and 12 for set CNJ02, linkage group 11 for set CNJ04, and linkage group 11 for set CNJ0x. For set CNJ02, linkage group 2 has a composite QTL at position 72.6cM for UBL+ULvW+UKEC+UKLvW, linkage group 3 has a year-stable metaQTL at position 55.8cM for TAcy, linkage group 10 has a compound QTL at position 29.4cM for UBM+UTBM+UMFM+MFM+UBW, linkage group 11 has three metaQTL at two positions, 29.4cM (UBM+UTBM+UMFM+MFM+UBW and UBL+ULvW+UKEC+UKLvW) and 38.4cM (UBM+UTBM+UMFM+MFM+UBW), and linkage group 12 has a collocated QTL at position 40.4cM for SFY+TY. Set CNJ04, linkage group 11 has a collocated QTL at position 10.3cM for UBL+ULvW. For set CNJ0x, linkage group 3 has a metaQTL at position 55.8cM for Tacy, linkage 11 has compound QTL at position 18.7cM for UBL+ULvW+UKEC+UKLvW, and linkage 12 has a composite QTL at position 40.4cM for SFY+TY.


```{r qtl-fig-cnj0x-grouped-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Noteworthy collocated QTL for population CNJ02 (A), CNJ04 (B), and both CNJ02 and CNJ04 (C), based on results from this study."}
knitr::include_graphics("../Data/publication/figures/cnj0x_qtl_collated.grouped.consensus.ABC.png")
```

#### Diaz-Garcia 2018b and CNJ0x

Figure \@ref(fig:qtl-fig-diazGarciaMassive-cnj0x-grouped-colocated) shows the primary QTL that compose the collocated QTL for sets Diaz-Garcia 2018b and CNJ0x & Diaz-Garcia 2018b, per Table 2.4. Together they represent stable QTL regions derived from QTL found in the current study for populations CNJ02, CNJ04, & GRYG. The traits that pass the filtering criteria outlined in Table 2.4 are chemistry, digital, and yield traits, with digital traits conveying berry color, a trait tightly associated with TAcy content (Diaz-Garcia et al. 2018b). Notable composite QTL are found on linkage groups 1, 3, 5, and 7 for set Diaz-Garcia 2018b and linkage group 3 for set CNJ0x & Diaz-Garcia 2018b. For set Diaz-Garcia 2018b, linkage group 1 has a composite QTL at position 31.7cM for October Brix, linkage group 3 has metaQTL at positions 8.8, 33.4, and 56.1cM for BCOLOR, BCOLOR+BCOLORVAR, and BCOLOR+TACY_SEP+BCOLORVAR+TACY_OCT, respectively. Linkage group 5 has a collocated QTL at position 47.7 for TA_OCT, linkage group 7 at position 52.2 for BRIX_SEP, linkage group 8 at position 14.7 for BRIX_OCT, linkage group 10 at 32.9cM for MFM_OCT, and linkage group 11 at positions 30.3 and 46.5 for MFM_SEP and BCOLOR+BCOLORVAR, respectively. Set CNJ0x & Diaz-Garcia 2018b has two composite QTL on linkage group 3 that are stable across this study and the Diaz-Garcia et al. 2018b QTL study. These metaQTL are found at positions 34.0cM and 56.1cM for traits BCOLOR+Tacy+BCOLORVAR and Tacy+BCOLOR+TACY_SEP+BCOLORVAR+TACY_OCT respectively.

```{r qtl-fig-diazGarciaMassive-cnj0x-grouped-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Noteworthy collocated QTL from Diaz-Garcia et al. 2018b in populations CNJ02, CNJ04, and GRYG (A). Cross study collocated QTL from Diaz-Garcia et al. 2018b and this study in populations CNJ02, CNJ04, and GRYG (B)."}
knitr::include_graphics("../Data/publication/figures/diazGarcia2018MassivePhenotyping_cnj0x.grouped.consensus.AB.pad.png")
```
#### Diaz-Garcia 2018a and CNJ0x

Figure \@ref(fig:qtl-fig-diazGarciaImage-cnj0x-grouped-colocated) shows the elemental QTL that compose the collocated QTL for sets Diaz-Garcia 2018a and CNJ0x & Diaz-Garcia 2018a, per Table 2.4. Together they represent stable QTL regions for populations CNJ02, CNJ04, & GRYG. The traits that pass the filtering criteria outlined in Table 2.4 are affiliated with yield and berry quality, with digital traits measuring berry shape and size parameters (Diaz-Garcia et al. 2018a). Prominent composite QTL are found on linkage groups 1, 2, 11, and 12 for set Diaz-Garcia 2018a and linkage groups 1, 2, 9, 11, and 12 for set CNJ0x & Diaz-Garcia 2018a. For set Diaz-Garcia 2018a, linkage group 1 has a composite QTL at around position 102cM for BW and BA, linkage group 2 has a metaQTL for LvW+EC at position 22.0cM, linkage group 11 has a composite QTL at 26.8cM for PH_PC2 and four metaQTL clustered at positions 76.4, 77.8, 80.0, and 82.4cM for PH_PC1, LvW+EC, BL, and BA respectively. Linkage group 12 has collocated QTL at position 43.5cM for LvW+EC and 54.9cM for PH_PC2.

```{r qtl-fig-diazGarciaImage-cnj0x-grouped-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Noteworthy collocated QTL from Diaz-Garcia et al. 2018a in population GRYG (A). Cross study collocated QTL from Diaz-Garcia et al. 2018a and this study in populations CNJ02, CNJ04, and GRYG (B)."}
knitr::include_graphics("../Data/publication/figures/diazGarcia2018ImagePhenotyping_cnj0x.grouped.consensus.AB.pad.png")
```

#### Schlautman 2015, Diaz-Garcia 2018b, and CNJ0x

Figure \@ref(fig:qtl-fig-schlautman-diazGarcia-cnj0x-grouped-colocated) illustrates the elemental QTL that compose the collocated QTL for sets Schlautman et al. 2015 and CNJ0x & Schlautman 2015 & Diaz-Garcia 2018a, per Table 2.4. These constitute stable QTL regions across three mapping studies from cranberry mapping populations CNJ02, CNJ04, & GRYG. Distinguished traits in figure 3.12 are associated with yield and berry quality. Notable composite QTL are found on linkage groups 6, 10, and 11 for set Schlautman et al. 2015 and linkage groups 1, 2, 9, 10, 11, and 12 for setCNJ0x & Schlautman 2015 & Diaz-Garcia 2018a. For set Schlautman et al. 2015, linkage group 6 has a compound QTL at around position 77cM for UTBM+MFM, linkage group 10 has two metaQTL for UBM+UMFM+MFM and UTBM+UBL+MFM at positions 33.3 and 42.0cM, linkage group 11 has four composite QTL; at 12.6cM for UBL+UMFM+MFM, at 18.7cM for PAC+SFY+TY, at 19.6cM for BBITY, and at 38.4cM for UBM+UTBM+UMFM+MFM+UBW. For set CNJ0x & Schlautman 2015 & Diaz-Garcia 2018a, linkage group 1 has a compound QTL at positions 101.6 for traits BA+UBW+BW+BL.  Linkage group 2 has a metaQTL at position 23.2cM for UBL+UMFM+BL. Linkage groups 9 and 10 each have a single metaQTL; linkage group 9 at 82.7cM for UBM+UTBM+BW, and linkage group 10 at 33.3cM for UBM+UMFM+MFM. Linkage group 11 has the highest number and density of metaQTL, with a cluster of 3 composite QTL from around 13cM to 21cM, one metaQTL at 38.4cM, and two collocated QTL at 77.8cM to 81.2cM. The first cluster has metaQTL at 13.1cM and 16.0cM for UBL+UMFM+MFM+BW+BL and EC+ULvW respectively, and metaQTL at 18.7cM and 21.1cM for PAC+SFY+TY and UKEC+LvW+ULvW. Linkage group 11 also has compound QTL at 38.4cM for UBM+UTBM+UMFM+MFM+UBW, at 77.8cM for LvW+EC+ULvW, and at 81.2cM for BA+BL+MFM. Linkage group 12 has two metaQTL for LvW+UKEC+EC, one at 45.0cM and one at 50.3cM.


```{r qtl-fig-schlautman-diazGarcia-cnj0x-grouped-colocated, echo=FALSE, message=FALSE, include=TRUE, results='asis', out.width="95%", fig.show="hold", fig.align="center", fig.cap="Noteworthy cross study collocated QTL from Schlautman et al. 2015 and this study in populations CNJ02 and CNJ04 (A). Cross study collocated QTL from Diaz-Garcia et al. 2018, Schlautman et al. 2015, and this study in populations CNJ02, CNJ04, and GRYG (B)."}
knitr::include_graphics("../Data/publication/figures/schlautman_diazGarciaImagePhenotyping_cnj0x.grouped.consensus.AB.pad.png")
```