# Discussion {#discuss}

## Resummarize importance of this study/analysis

A recent buildup of the quantity and quality of cranberry molecular resources has the potential to propel a renaissance in cranberry breeding.  
Advances in the genetic capital of this important fruit crop include development of high density linkage maps,
construction of mitochondrial and nuclear genome assemblies, a plethora of QTL mapping studies in a variety of important traits,
and an early feasibility study of genomic prediction and genomic selection (TODO: references).  

Shifts and innovations in cranberry phenotyping methods have paralleled advances in genetic resource development.
Up until 60 years ago, breeders would select cranberry breeding material based on traits measured from the fundamental 
unit of cranberry productivity: the reproductive upright.  Since then, modernization of farming technology and
management methods, a better understanding cranberry physiology, scaling up planting of more productive cultivars, 
climate change, and evolving breeding objectives has led (transofrmed) to growers and breeders needing to apply newer trait assessment 
metrics and methods in order maintain the pace of germplasm advancement.  This shift has predominantly involved moving away from assessing 
cranberry uprights as a fundamental measure of cultivar caliber to more holistic measurements at a plot-level.
gap in our understanding b/w the two different methodologies ...
This study offers a unique opportunity to assess the genetic basis of reproductive upright traits that were 
classically collected by breeders in order to advance cranberry domestication.
offers the chance to compare old to new and validate that newer methods are sound.


 - This QTL mapping study offers a unique perspective on ...
 - Proof of concept comparing older studies to newer ones.
 - Better understand trait correlations in two third-generation breeding populations
 - 


## Summarize results

## High level comparison of these results to other previous studies in cranberries
    - previous QTL mapping of FRR QTLs done in very small populations [@Georgi2013] - since the linkage group numbers
          - CNJ04 population shares a parent 'Stevens' with many of the crosses made in [@Georgi2013]
          - Using listed markers near QTLs from this paper, was able to remap and show if QTLs overlap at all
            - One interesting overlap of Georgi is MRR, marker scf3072b (lg11@26cM on our map) with all-year PFR on lg11@24cM
            - A closely linked one was also found in CNJ04 found on lg11@17cM
    - Schlautman2015 paper
        - MFM - Closest overlap with QTL on LG10@48.52 (LG@37,LG@39 for UMFM and MFM, respectively).  Also QTL on LG11,
          but not really closely located to the ones in Schlautman's paper
        - TY - LG4@88.99 -> LG4@85, LG6@56.80 -> LG6@52, LG11@24 -> LG11@20, although in our paper, our largest TY QTL
          was on LG12
        - BBI-TY - LG11@24 -> LG11@20
        - h2 of TY 64 => .741
        - h2 of MFM => .7 -> 0.81
        - higher heritability on my end may be due to usage of BLUPs to calculate genomic heritability
    - Diaz 2018 paper - QTLs with chemistry and some digital image phenotyping
        - 
    - Diaz 2018 paper - image phenotyping
        - Given that this is an entirely different population, it is difficult to really compare results
        - However, our 'largest berry width' has one closely associated overlap with Luis' collocated QTL on LG1
            LG1@103.2 -> LG1@100
    - heritability of upright traits and how they compared to Vorsa's?
    - mid-level progeny FRR revealed high heritability of 0.81 [@JohnsonCicaleseFRR2015]
        - I found similarly high heritability in CNJ02 population (h2=0.73)
        - lower heritability for FRR found in CNJ04 (h2=0.46), likely due to downward biasing in smaller population size
          when calculating BLUPs (need reference here)
    - steal page from Luis' 'massive phenotyping' paper and mention how results are consistent with previous Vorsa
      studies
    - how results overlap with 

## Limitations and where to go from here.

- Creating a central site to unify previous studies in order to help propel breeding forward
### how digital ag. can help upend/reveal more, and take this a step further


## Take home message


For CNJ02 population:

Tacy and PAC inversely correlated with SFY and positively correlated with PFR could mean that fruit rot is undesirably 
linked with higher levels of Tacy/PAC (desirable traits), or could simply be an indicator that these genotypes undergo 
berry development and ripening sooner than the others (since everything harvested at the same time).  A few additional
indicators of this relationship between sound fruit yield and Tacy/PAC can be found in the modestly negative
correlations of *SFY × PAC* ($\rho =  `r p.cnj02.c['SFY','PAC']`$), *SFY × Tacy* ($\rho =  `r p.cnj02.c['SFY','Tacy']`$), 
*UNAB × PAC* ($\rho =  `r p.cnj02.c['UNAB','PAC']`$), *MFM × PAC* ($\rho =  `r p.cnj02.c['MFM','PAC']`$), 
*MFM × UNAF* ($\rho =  `r p.cnj02.c['MFM','UNAF']`$), and *MFM × UBBL* ($\rho =  `r p.cnj02.c['MFM','UBBL']`$).

Compared to CNJ02, CNJ04 has  weaker correlations for traits
*UBM x UBW* (`r p.cnj04.c['UBM','UBW']` vs. `r p.cnj02.c['UBM','UBW']`) and *UBM × UBL* (`r p.cnj04.c['UBM','UBL']` vs. `r p.cnj02.c['UBM','UBL']`). - Due to less variation/segregation in width and length 
in berries found in CNJ04?  Look at BLUP and trait summary stats tables

Comment and conjecture on noteworthy correlations discussed in results:

For CNJ04, there appears to be a modest negative correlation between berry size and Brix.  This could represent some
form of linkage drag in this population, as seen in other progeny derived from a Steven's parent (Sweetie, progeny of a
cross of Stevens × Ben Lear).

No single large *TY* QTLs --  seem to be spread out across the genome - consistent with yield being a quantitative trait
with a lot of genetic players involved.


*UCLP x UCLS* is interesting in that calyx lobe fold pattern could have more to do developmentally with calyx lobe size
(pleiotropy), or the genes are strongly linked.

It would be interesting to compare differential fruit coloring/Tacy with differential fruit rot (check out Luis' paper)
-- this could allow us to decouple ripening stage from negatively linked sound fruit yield with Tacy/PAC levels.

##Heritabilities

High heritability of berry length in both populations would seem to indicate that this would be a good way of selecting
for larger berries.  Even higher heritabilities in berry length vs. width indicates that this is a great trait to breed
for when selecting for particular berry shape characteristics important for processing (rounder berries preferred).  If
it is possible to select genotypes with larger berry length and low LvW, along with low rot rates, this could be a good
way to produce genotypes with good processing abilities.

Evidence of very low heritabilities in traits such as *Brix* and *TA* in population CNJ02 over CNJ04 could simply
indicate less genetic variation for these traits in CNJ02, but the baseline could be higher -- show a figure of 
inter-population BLUPS side-by-side to get a good perspective on the absolute amount of selectable additive genetic variation 
b/w populations.

##QTLs

- Make a comment that smaller population size of CNJ04 could result in overfitting, and hence the higher average model
  fits than in CNJ02 (almost 17% higher!)
- CNJ02 - Mention that the largest set of collocated QTLs on LG 3 is intriguing, b/c one of them is derived from an entirely different set of data (plot data -- plot mean fruit weigth) versus the upright mean fruit weight data.  Shows overlap b/w datasets (check to see how UMFM and MFM correlate) - Mention that total yield also has a closely linked QTL on LG3, but not collocated likely b/c total yield involves a more complex mix of factors.
    - On linkage group 11, largest *berry weight* and *upright mean fruit mass* are collocated at 30cM, while plot mean fruit mass is closely
linked at 25cM. All three have the same genotype (*BD*) with the largest effect, indicating that they could all be
driven by the same QTL. 
    - Close linkage of PFR and Tacy on LG7, with shared patterns in genotype effects, could also be associated with
      early-ripening varieties.  Compare to the differential QTLs found in Luis' paper.

- For CNJ04, although a lower LvW ratio is preferred for rounder berries, selecting for this could result in smaller
  berries by dragging in berries with shorter lengths.  Thus, it would make sense to select for berries with larger
  widths to offset this trend.  Although the heritability of berry width is such that it is not displayed in table
  \@ref(tab:qtl-table-cnj04-sw) because its heritability (0.48) is just below the cutoff of 0.5, lg 9 pos 64.23cM shows
  a modestly positive effect for genotype *AD*.  Since it is on a separate linkage group from the primary QTL for upright *berry
  length* (lg 11), it is easy to select for berries with larger widths and lower lengths without worrying about linkage
  drag.  Also note that CNJ04 much more explicit variation in berry length compared to width, so it is more difficult
  to select for rounder berries without negatively impacting berry size, and thus yield.  Upright *total berry
weight* does not even share its top 2 QTL with the same linkage groups as *total yield*, but given that the intertrait
correlation between them in insignificant, it is not surprising that they do not share QTLs.
- Mention that despite having some modest correlations b/w chemistry traits and chemistry/yield traits, that a lack of
  overlap in top QTLs makes it difficult to pinpoint a genetic basis of this, if any.  This could be due to the small
  population size being insufficient to unravel any shared/linked QTLs here, or ...

- Both
  - really emphasize the shared QTL on lg 12 highlighting its importance in controlling total yield.

- Epistatic Interactions
      - emphasize that the top epistatic QTL are on separate linkage groups -- not surprising given that the smaller
        population sizes and lower recombination history gives less chance for neighboring QTL to break their linkages
        and give the appropriate number of samples of all pairwise combinations for determining significance.
      - not surprising that CNJ04 has a lot less epistatic QTL than CNJ02, given that it is less than half the
        population size of CNJ02.
      - mention that a changing in effect rank, or ordering per genotype across it's interacting pair also indicates
        that an interaction effect exists.

- Talk about how translating to biallelic markers only and using population-specific consensus maps could possibly
  help some with increasing power to detect QTL and helping interpret effects more meaningfully.
